// @copilotkit/runtime-client-gql/dist/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_graphql = require('./graphql/@generated/graphql.cjs');
const require_CopilotRuntimeClient = require('./client/CopilotRuntimeClient.cjs');
const require_types = require('./client/types.cjs');
const require_conversion = require('./client/conversion.cjs');
const require_agui_to_gql = require('./message-conversion/agui-to-gql.cjs');
const require_gql_to_agui = require('./message-conversion/gql-to-agui.cjs');

exports.ActionExecutionMessage = require_types.ActionExecutionMessage;
exports.ActionInputAvailability = require_graphql.ActionInputAvailability;
exports.AgentStateMessage = require_types.AgentStateMessage;
exports.AvailableAgentsDocument = require_graphql.AvailableAgentsDocument;
exports.CopilotRequestType = require_graphql.CopilotRequestType;
exports.CopilotRuntimeClient = require_CopilotRuntimeClient.CopilotRuntimeClient;
exports.FailedResponseStatusReason = require_graphql.FailedResponseStatusReason;
exports.GenerateCopilotResponseDocument = require_graphql.GenerateCopilotResponseDocument;
exports.ImageMessage = require_types.ImageMessage;
exports.LoadAgentStateDocument = require_graphql.LoadAgentStateDocument;
exports.Message = require_types.Message;
exports.MessageRole = require_graphql.MessageRole;
exports.MessageStatusCode = require_graphql.MessageStatusCode;
exports.MetaEventName = require_graphql.MetaEventName;
exports.ResponseStatusCode = require_graphql.ResponseStatusCode;
exports.ResultMessage = require_types.ResultMessage;
exports.Role = require_types.Role;
exports.TextMessage = require_types.TextMessage;
exports.aguiMessageWithImageToGQLMessage = require_agui_to_gql.aguiMessageWithImageToGQLMessage;
exports.aguiMessageWithRenderToGQL = require_agui_to_gql.aguiMessageWithRenderToGQL;
exports.aguiTextMessageToGQLMessage = require_agui_to_gql.aguiTextMessageToGQLMessage;
exports.aguiToGQL = require_agui_to_gql.aguiToGQL;
exports.aguiToolCallToGQLActionExecution = require_agui_to_gql.aguiToolCallToGQLActionExecution;
exports.aguiToolMessageToGQLResultMessage = require_agui_to_gql.aguiToolMessageToGQLResultMessage;
exports.convertGqlOutputToMessages = require_conversion.convertGqlOutputToMessages;
exports.convertMessagesToGqlInput = require_conversion.convertMessagesToGqlInput;
exports.filterAdjacentAgentStateMessages = require_conversion.filterAdjacentAgentStateMessages;
exports.filterAgentStateMessages = require_conversion.filterAgentStateMessages;
exports.gqlActionExecutionMessageToAGUIMessage = require_gql_to_agui.gqlActionExecutionMessageToAGUIMessage;
exports.gqlImageMessageToAGUIMessage = require_gql_to_agui.gqlImageMessageToAGUIMessage;
exports.gqlResultMessageToAGUIMessage = require_gql_to_agui.gqlResultMessageToAGUIMessage;
exports.gqlTextMessageToAGUIMessage = require_gql_to_agui.gqlTextMessageToAGUIMessage;
exports.gqlToAGUI = require_gql_to_agui.gqlToAGUI;
exports.langGraphInterruptEvent = require_types.langGraphInterruptEvent;
exports.loadMessagesFromJsonRepresentation = require_conversion.loadMessagesFromJsonRepresentation;
