// @copilotkit/shared/dist/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_clipboard = require('./utils/clipboard.cjs');
const require_conditions = require('./utils/conditions.cjs');
const require_console_styling = require('./utils/console-styling.cjs');
const require_errors = require('./utils/errors.cjs');
const require_json_schema = require('./utils/json-schema.cjs');
const require_types = require('./utils/types.cjs');
const require_random_id = require('./utils/random-id.cjs');
const require_requests = require('./utils/requests.cjs');
const require_index = require('./utils/index.cjs');
const require_index$1 = require('./constants/index.cjs');
const require_package = require('./package.cjs');
const require_telemetry_client = require('./telemetry/telemetry-client.cjs');
const require_debug = require('./debug.cjs');
const require_standard_schema = require('./standard-schema.cjs');
const require_utils = require('./attachments/utils.cjs');
const require_logger = require('./logger.cjs');
const require_finalize_events = require('./finalize-events.cjs');
const require_transcription_errors = require('./transcription-errors.cjs');
const require_a2ui_prompts = require('./a2ui-prompts.cjs');

//#region src/index.ts
const COPILOTKIT_VERSION = require_package.version;
/**
* Client-safe license context factory.
*
* When status is null (no token provided), all features return true
* (unlicensed = unrestricted, with branding). This is inlined here to
* avoid importing the full license-verifier bundle (which depends on
* Node's `crypto`) into browser bundles.
*/
function createLicenseContextValue(status) {
	return {
		status: null,
		license: null,
		checkFeature: () => true,
		getLimit: () => null
	};
}

//#endregion
exports.A2UI_DEFAULT_DESIGN_GUIDELINES = require_a2ui_prompts.A2UI_DEFAULT_DESIGN_GUIDELINES;
exports.A2UI_DEFAULT_GENERATION_GUIDELINES = require_a2ui_prompts.A2UI_DEFAULT_GENERATION_GUIDELINES;
exports.AG_UI_CHANNEL_EVENT = require_index$1.AG_UI_CHANNEL_EVENT;
exports.BANNER_ERROR_NAMES = require_errors.BANNER_ERROR_NAMES;
exports.COPILOTKIT_VERSION = COPILOTKIT_VERSION;
exports.COPILOT_CLOUD_API_URL = require_index$1.COPILOT_CLOUD_API_URL;
exports.COPILOT_CLOUD_CHAT_URL = require_index$1.COPILOT_CLOUD_CHAT_URL;
exports.COPILOT_CLOUD_ERROR_NAMES = require_errors.COPILOT_CLOUD_ERROR_NAMES;
exports.COPILOT_CLOUD_PUBLIC_API_KEY_HEADER = require_index$1.COPILOT_CLOUD_PUBLIC_API_KEY_HEADER;
exports.COPILOT_CLOUD_VERSION = require_index$1.COPILOT_CLOUD_VERSION;
exports.ConfigurationError = require_errors.ConfigurationError;
exports.ConsoleColors = require_console_styling.ConsoleColors;
exports.ConsoleStyles = require_console_styling.ConsoleStyles;
exports.CopilotKitAgentDiscoveryError = require_errors.CopilotKitAgentDiscoveryError;
exports.CopilotKitApiDiscoveryError = require_errors.CopilotKitApiDiscoveryError;
exports.CopilotKitError = require_errors.CopilotKitError;
exports.CopilotKitErrorCode = require_errors.CopilotKitErrorCode;
exports.CopilotKitLowLevelError = require_errors.CopilotKitLowLevelError;
exports.CopilotKitMisuseError = require_errors.CopilotKitMisuseError;
exports.CopilotKitRemoteEndpointDiscoveryError = require_errors.CopilotKitRemoteEndpointDiscoveryError;
exports.CopilotKitVersionMismatchError = require_errors.CopilotKitVersionMismatchError;
exports.DEFAULT_AGENT_ID = require_index$1.DEFAULT_AGENT_ID;
exports.ERROR_CONFIG = require_errors.ERROR_CONFIG;
exports.ERROR_NAMES = require_errors.ERROR_NAMES;
exports.ErrorVisibility = require_errors.ErrorVisibility;
exports.MissingPublicApiKeyError = require_errors.MissingPublicApiKeyError;
exports.RUNTIME_MODE_INTELLIGENCE = require_types.RUNTIME_MODE_INTELLIGENCE;
exports.RUNTIME_MODE_SSE = require_types.RUNTIME_MODE_SSE;
exports.ResolvedCopilotKitError = require_errors.ResolvedCopilotKitError;
exports.Severity = require_errors.Severity;
exports.TelemetryClient = require_telemetry_client.TelemetryClient;
exports.TranscriptionErrorCode = require_transcription_errors.TranscriptionErrorCode;
exports.TranscriptionErrors = require_transcription_errors.TranscriptionErrors;
exports.UpgradeRequiredError = require_errors.UpgradeRequiredError;
exports.actionParametersToJsonSchema = require_json_schema.actionParametersToJsonSchema;
exports.convertJsonSchemaToZodSchema = require_json_schema.convertJsonSchemaToZodSchema;
exports.copyToClipboard = require_clipboard.copyToClipboard;
exports.createLicenseContextValue = createLicenseContextValue;
exports.dataToUUID = require_random_id.dataToUUID;
exports.ensureStructuredError = require_errors.ensureStructuredError;
exports.exceedsMaxSize = require_utils.exceedsMaxSize;
exports.executeConditions = require_conditions.executeConditions;
exports.finalizeRunEvents = require_finalize_events.finalizeRunEvents;
exports.formatFileSize = require_utils.formatFileSize;
exports.generateVideoThumbnail = require_utils.generateVideoThumbnail;
exports.getDocumentIcon = require_utils.getDocumentIcon;
exports.getModalityFromMimeType = require_utils.getModalityFromMimeType;
exports.getPossibleVersionMismatch = require_errors.getPossibleVersionMismatch;
exports.getSourceUrl = require_utils.getSourceUrl;
exports.getZodParameters = require_json_schema.getZodParameters;
exports.isMacOS = require_index.isMacOS;
exports.isStructuredCopilotKitError = require_errors.isStructuredCopilotKitError;
exports.isTelemetryDisabled = require_telemetry_client.isTelemetryDisabled;
exports.isValidUUID = require_random_id.isValidUUID;
exports.jsonSchemaToActionParameters = require_json_schema.jsonSchemaToActionParameters;
exports.logCopilotKitPlatformMessage = require_console_styling.logCopilotKitPlatformMessage;
exports.logStyled = require_console_styling.logStyled;
exports.logger = require_logger.logger;
exports.matchesAcceptFilter = require_utils.matchesAcceptFilter;
exports.parseJson = require_index.parseJson;
exports.partialJSONParse = require_index.partialJSONParse;
exports.phoenixExponentialBackoff = require_index.phoenixExponentialBackoff;
exports.publicApiKeyRequired = require_console_styling.publicApiKeyRequired;
exports.randomId = require_random_id.randomId;
exports.randomUUID = require_random_id.randomUUID;
exports.readBody = require_requests.readBody;
exports.readFileAsBase64 = require_utils.readFileAsBase64;
exports.resolveDebugConfig = require_debug.resolveDebugConfig;
exports.safeParseToolArgs = require_index.safeParseToolArgs;
exports.schemaToJsonSchema = require_standard_schema.schemaToJsonSchema;
exports.styledConsole = require_console_styling.styledConsole;
exports.tryMap = require_index.tryMap;
