// @copilotkit/shared/dist/index.umd.js
(function(global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ?  factory(exports, require('graphql'), require('zod'), require('uuid'), require('partial-json'), require('@segment/analytics-node'), require('chalk'), require('@ag-ui/client')) :
  typeof define === 'function' && define.amd ? define(['exports', 'graphql', 'zod', 'uuid', 'partial-json', '@segment/analytics-node', 'chalk', '@ag-ui/client'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.CopilotKitShared = {}), global.GraphQL,global.Zod,global.UUID,global.PartialJSON,global.SegmentAnalyticsNode,global.chalk,global.AgUIClient));
})(this, function(exports, graphql, zod, uuid, partial_json, _segment_analytics_node, chalk, _ag_ui_client) {
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
//#region \0rolldown/runtime.js
	var __create = Object.create;
	var __defProp = Object.defineProperty;
	var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
	var __getOwnPropNames = Object.getOwnPropertyNames;
	var __getProtoOf = Object.getPrototypeOf;
	var __hasOwnProp = Object.prototype.hasOwnProperty;
	var __copyProps = (to, from, except, desc) => {
		if (from && typeof from === "object" || typeof from === "function") {
			for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
				key = keys[i];
				if (!__hasOwnProp.call(to, key) && key !== except) {
					__defProp(to, key, {
						get: ((k) => from[k]).bind(null, key),
						enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
					});
				}
			}
		}
		return to;
	};
	var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
		value: mod,
		enumerable: true
	}) : target, mod));

//#endregion
partial_json = __toESM(partial_json);

//#region src/utils/clipboard.ts
/**
	* Safely copies text to the clipboard.
	*
	* Checks that the Clipboard API is available before attempting the write,
	* and catches any errors (e.g. permission denied, insecure context).
	*
	* @param text - The text to copy to the clipboard.
	* @returns `true` if the text was successfully copied, `false` otherwise.
	*/
	async function copyToClipboard(text) {
		var _navigator$clipboard;
		if (!((_navigator$clipboard = navigator.clipboard) === null || _navigator$clipboard === void 0 ? void 0 : _navigator$clipboard.writeText)) {
			console.error("Clipboard API is not available");
			return false;
		}
		try {
			await navigator.clipboard.writeText(text);
			return true;
		} catch (err) {
			console.error("Failed to copy to clipboard:", err);
			return false;
		}
	}

//#endregion
//#region src/utils/conditions.ts
	function executeConditions({ conditions, value }) {
		if (!(conditions === null || conditions === void 0 ? void 0 : conditions.length)) return true;
		return conditions.every((condition) => executeCondition(condition, value));
	}
	function executeCondition(condition, value) {
		const targetValue = condition.path ? getValueFromPath(value, condition.path) : value;
		switch (condition.rule) {
			case "AND": return condition.conditions.every((c) => executeCondition(c, value));
			case "OR": return condition.conditions.some((c) => executeCondition(c, value));
			case "NOT": return !condition.conditions.every((c) => executeCondition(c, value));
			case "EQUALS": return targetValue === condition.value;
			case "NOT_EQUALS": return targetValue !== condition.value;
			case "GREATER_THAN": return targetValue > condition.value;
			case "LESS_THAN": return targetValue < condition.value;
			case "CONTAINS": return Array.isArray(targetValue) && targetValue.includes(condition.value);
			case "NOT_CONTAINS": return Array.isArray(targetValue) && !targetValue.includes(condition.value);
			case "MATCHES": return new RegExp(condition.value).test(String(targetValue));
			case "STARTS_WITH": return String(targetValue).startsWith(condition.value);
			case "ENDS_WITH": return String(targetValue).endsWith(condition.value);
			case "EXISTS": return targetValue !== void 0 && targetValue !== null;
			case "NOT_EXISTS": return targetValue === void 0 || targetValue === null;
		}
	}
	function getValueFromPath(obj, path) {
		return path.split(".").reduce((acc, part) => acc === null || acc === void 0 ? void 0 : acc[part], obj);
	}

//#endregion
//#region src/utils/console-styling.ts
/**
	* Console styling utilities for CopilotKit branded messages
	* Provides consistent, readable colors across light and dark console themes
	*/
	/**
	* Color palette optimized for console readability
	*/
	const ConsoleColors = {
		primary: "#007acc",
		success: "#22c55e",
		feature: "#a855f7",
		cta: "#ef4444",
		info: "#06b6d4",
		inherit: "inherit",
		warning: "#f59e0b"
	};
	/**
	* Console style templates for common patterns
	*/
	const ConsoleStyles = {
		header: `color: ${ConsoleColors.warning}; font-weight: bold; font-size: 16px;`,
		section: `color: ${ConsoleColors.success}; font-weight: bold;`,
		highlight: `color: ${ConsoleColors.feature}; font-weight: bold;`,
		cta: `color: ${ConsoleColors.success}; font-weight: bold;`,
		info: `color: ${ConsoleColors.info}; font-weight: bold;`,
		link: `color: ${ConsoleColors.primary}; text-decoration: underline;`,
		body: `color: ${ConsoleColors.inherit};`,
		warning: `color: ${ConsoleColors.cta}; font-weight: bold;`
	};
	/**
	* Styled console message for CopilotKit Platform promotion
	* Displays a beautiful, branded advertisement in the console
	*/
	function logCopilotKitPlatformMessage() {
		console.log(`%cCopilotKit Warning%c

useCopilotChatHeadless_c provides full compatibility with CopilotKit's newly released Headless UI feature set. To enable this premium feature, add your public license key, available for free at:

%chttps://cloud.copilotkit.ai%c

Alternatively, useCopilotChat is available for basic programmatic control, and does not require an API key.

To learn more about premium features, read the documentation here:

%chttps://docs.copilotkit.ai/premium%c`, ConsoleStyles.header, ConsoleStyles.body, ConsoleStyles.cta, ConsoleStyles.body, ConsoleStyles.link, ConsoleStyles.body);
	}
	function publicApiKeyRequired(feature) {
		console.log(`
%cCopilotKit Warning%c \n
In order to use ${feature}, you need to add your CopilotKit API key, available for free at https://cloud.copilotkit.ai.
    `.trim(), ConsoleStyles.header, ConsoleStyles.body);
	}
	/**
	* Create a styled console message with custom content
	*
	* @param template - Template string with %c placeholders
	* @param styles - Array of style strings matching the %c placeholders
	*
	* @example
	* ```typescript
	* logStyled(
	*   '%cCopilotKit%c Welcome to the platform!',
	*   [ConsoleStyles.header, ConsoleStyles.body]
	* );
	* ```
	*/
	function logStyled(template, styles) {
		console.log(template, ...styles);
	}
	/**
	* Quick styled console methods for common use cases
	*/
	const styledConsole = {
		success: (message) => logStyled(`%c✅ ${message}`, [ConsoleStyles.section]),
		info: (message) => logStyled(`%cℹ️ ${message}`, [ConsoleStyles.info]),
		feature: (message) => logStyled(`%c✨ ${message}`, [ConsoleStyles.highlight]),
		cta: (message) => logStyled(`%c🚀 ${message}`, [ConsoleStyles.cta]),
		logCopilotKitPlatformMessage,
		publicApiKeyRequired
	};

//#endregion
//#region src/utils/errors.ts
	let Severity = /* @__PURE__ */ function(Severity) {
		Severity["CRITICAL"] = "critical";
		Severity["WARNING"] = "warning";
		Severity["INFO"] = "info";
		return Severity;
	}({});
	let ErrorVisibility = /* @__PURE__ */ function(ErrorVisibility) {
		ErrorVisibility["BANNER"] = "banner";
		ErrorVisibility["TOAST"] = "toast";
		ErrorVisibility["SILENT"] = "silent";
		ErrorVisibility["DEV_ONLY"] = "dev_only";
		return ErrorVisibility;
	}({});
	const ERROR_NAMES = {
		COPILOT_ERROR: "CopilotError",
		COPILOT_API_DISCOVERY_ERROR: "CopilotApiDiscoveryError",
		COPILOT_REMOTE_ENDPOINT_DISCOVERY_ERROR: "CopilotKitRemoteEndpointDiscoveryError",
		COPILOT_KIT_AGENT_DISCOVERY_ERROR: "CopilotKitAgentDiscoveryError",
		COPILOT_KIT_LOW_LEVEL_ERROR: "CopilotKitLowLevelError",
		COPILOT_KIT_VERSION_MISMATCH_ERROR: "CopilotKitVersionMismatchError",
		RESOLVED_COPILOT_KIT_ERROR: "ResolvedCopilotKitError",
		CONFIGURATION_ERROR: "ConfigurationError",
		MISSING_PUBLIC_API_KEY_ERROR: "MissingPublicApiKeyError",
		UPGRADE_REQUIRED_ERROR: "UpgradeRequiredError"
	};
	const BANNER_ERROR_NAMES = [
		ERROR_NAMES.CONFIGURATION_ERROR,
		ERROR_NAMES.MISSING_PUBLIC_API_KEY_ERROR,
		ERROR_NAMES.UPGRADE_REQUIRED_ERROR,
		ERROR_NAMES.COPILOT_API_DISCOVERY_ERROR,
		ERROR_NAMES.COPILOT_REMOTE_ENDPOINT_DISCOVERY_ERROR,
		ERROR_NAMES.COPILOT_KIT_AGENT_DISCOVERY_ERROR
	];
	const COPILOT_CLOUD_ERROR_NAMES = BANNER_ERROR_NAMES;
	let CopilotKitErrorCode = /* @__PURE__ */ function(CopilotKitErrorCode) {
		CopilotKitErrorCode["NETWORK_ERROR"] = "NETWORK_ERROR";
		CopilotKitErrorCode["NOT_FOUND"] = "NOT_FOUND";
		CopilotKitErrorCode["AGENT_NOT_FOUND"] = "AGENT_NOT_FOUND";
		CopilotKitErrorCode["API_NOT_FOUND"] = "API_NOT_FOUND";
		CopilotKitErrorCode["REMOTE_ENDPOINT_NOT_FOUND"] = "REMOTE_ENDPOINT_NOT_FOUND";
		CopilotKitErrorCode["AUTHENTICATION_ERROR"] = "AUTHENTICATION_ERROR";
		CopilotKitErrorCode["MISUSE"] = "MISUSE";
		CopilotKitErrorCode["UNKNOWN"] = "UNKNOWN";
		CopilotKitErrorCode["VERSION_MISMATCH"] = "VERSION_MISMATCH";
		CopilotKitErrorCode["CONFIGURATION_ERROR"] = "CONFIGURATION_ERROR";
		CopilotKitErrorCode["MISSING_PUBLIC_API_KEY_ERROR"] = "MISSING_PUBLIC_API_KEY_ERROR";
		CopilotKitErrorCode["UPGRADE_REQUIRED_ERROR"] = "UPGRADE_REQUIRED_ERROR";
		return CopilotKitErrorCode;
	}({});
	const BASE_URL = "https://docs.copilotkit.ai";
	const getSeeMoreMarkdown = (link) => `See more: [${link}](${link})`;
	const ERROR_CONFIG = {
		[CopilotKitErrorCode.NETWORK_ERROR]: {
			statusCode: 503,
			troubleshootingUrl: `${BASE_URL}/troubleshooting/common-issues#i-am-getting-a-network-errors--api-not-found`,
			visibility: ErrorVisibility.BANNER,
			severity: Severity.CRITICAL
		},
		[CopilotKitErrorCode.NOT_FOUND]: {
			statusCode: 404,
			troubleshootingUrl: `${BASE_URL}/troubleshooting/common-issues#i-am-getting-a-network-errors--api-not-found`,
			visibility: ErrorVisibility.BANNER,
			severity: Severity.CRITICAL
		},
		[CopilotKitErrorCode.AGENT_NOT_FOUND]: {
			statusCode: 500,
			troubleshootingUrl: `${BASE_URL}/coagents/troubleshooting/common-issues#i-am-getting-agent-not-found-error`,
			visibility: ErrorVisibility.BANNER,
			severity: Severity.CRITICAL
		},
		[CopilotKitErrorCode.API_NOT_FOUND]: {
			statusCode: 404,
			troubleshootingUrl: `${BASE_URL}/troubleshooting/common-issues#i-am-getting-a-network-errors--api-not-found`,
			visibility: ErrorVisibility.BANNER,
			severity: Severity.CRITICAL
		},
		[CopilotKitErrorCode.REMOTE_ENDPOINT_NOT_FOUND]: {
			statusCode: 404,
			troubleshootingUrl: `${BASE_URL}/troubleshooting/common-issues#i-am-getting-copilotkits-remote-endpoint-not-found-error`,
			visibility: ErrorVisibility.BANNER,
			severity: Severity.CRITICAL
		},
		[CopilotKitErrorCode.AUTHENTICATION_ERROR]: {
			statusCode: 401,
			troubleshootingUrl: `${BASE_URL}/troubleshooting/common-issues#authentication-errors`,
			visibility: ErrorVisibility.BANNER,
			severity: Severity.CRITICAL
		},
		[CopilotKitErrorCode.MISUSE]: {
			statusCode: 400,
			troubleshootingUrl: null,
			visibility: ErrorVisibility.DEV_ONLY,
			severity: Severity.WARNING
		},
		[CopilotKitErrorCode.UNKNOWN]: {
			statusCode: 500,
			visibility: ErrorVisibility.TOAST,
			severity: Severity.CRITICAL
		},
		[CopilotKitErrorCode.CONFIGURATION_ERROR]: {
			statusCode: 400,
			troubleshootingUrl: null,
			severity: Severity.WARNING,
			visibility: ErrorVisibility.BANNER
		},
		[CopilotKitErrorCode.MISSING_PUBLIC_API_KEY_ERROR]: {
			statusCode: 400,
			troubleshootingUrl: null,
			severity: Severity.CRITICAL,
			visibility: ErrorVisibility.BANNER
		},
		[CopilotKitErrorCode.UPGRADE_REQUIRED_ERROR]: {
			statusCode: 402,
			troubleshootingUrl: null,
			severity: Severity.WARNING,
			visibility: ErrorVisibility.BANNER
		},
		[CopilotKitErrorCode.VERSION_MISMATCH]: {
			statusCode: 400,
			troubleshootingUrl: null,
			visibility: ErrorVisibility.DEV_ONLY,
			severity: Severity.INFO
		}
	};
	var CopilotKitError = class extends graphql.GraphQLError {
		constructor({ message = "Unknown error occurred", code, severity, visibility }) {
			var _ref;
			const name = ERROR_NAMES.COPILOT_ERROR;
			const config = ERROR_CONFIG[code];
			const { statusCode } = config;
			const resolvedVisibility = (_ref = visibility !== null && visibility !== void 0 ? visibility : config.visibility) !== null && _ref !== void 0 ? _ref : ErrorVisibility.TOAST;
			const resolvedSeverity = severity !== null && severity !== void 0 ? severity : "severity" in config ? config.severity : void 0;
			super(message, { extensions: {
				name,
				statusCode,
				code,
				visibility: resolvedVisibility,
				severity: resolvedSeverity,
				troubleshootingUrl: "troubleshootingUrl" in config ? config.troubleshootingUrl : null,
				originalError: {
					message,
					stack: (/* @__PURE__ */ new Error()).stack
				}
			} });
			this.code = code;
			this.name = name;
			this.statusCode = statusCode;
			this.severity = resolvedSeverity;
			this.visibility = resolvedVisibility;
		}
	};
	/**
	* Error thrown when we can identify wrong usage of our components.
	* This helps us notify the developer before real errors can happen
	*
	* @extends CopilotKitError
	*/
	var CopilotKitMisuseError = class extends CopilotKitError {
		constructor({ message, code = CopilotKitErrorCode.MISUSE }) {
			const docsLink = "troubleshootingUrl" in ERROR_CONFIG[code] && ERROR_CONFIG[code].troubleshootingUrl ? getSeeMoreMarkdown(ERROR_CONFIG[code].troubleshootingUrl) : null;
			const finalMessage = docsLink ? `${message}.\n\n${docsLink}` : message;
			super({
				message: finalMessage,
				code
			});
			this.name = ERROR_NAMES.COPILOT_API_DISCOVERY_ERROR;
		}
	};
	const getVersionMismatchErrorMessage = ({ reactCoreVersion, runtimeVersion, runtimeClientGqlVersion }) => `Version mismatch detected: @copilotkit/runtime@${runtimeVersion !== null && runtimeVersion !== void 0 ? runtimeVersion : ""} is not compatible with @copilotkit/react-core@${reactCoreVersion} and @copilotkit/runtime-client-gql@${runtimeClientGqlVersion}. Please ensure all installed copilotkit packages are on the same version.`;
	/**
	* Error thrown when CPK versions does not match
	*
	* @extends CopilotKitError
	*/
	var CopilotKitVersionMismatchError = class extends CopilotKitError {
		constructor({ reactCoreVersion, runtimeVersion, runtimeClientGqlVersion }) {
			const code = CopilotKitErrorCode.VERSION_MISMATCH;
			super({
				message: getVersionMismatchErrorMessage({
					reactCoreVersion,
					runtimeVersion,
					runtimeClientGqlVersion
				}),
				code
			});
			this.name = ERROR_NAMES.COPILOT_KIT_VERSION_MISMATCH_ERROR;
		}
	};
	/**
	* Error thrown when the CopilotKit API endpoint cannot be discovered or accessed.
	* This typically occurs when:
	* - The API endpoint URL is invalid or misconfigured
	* - The API service is not running at the expected location
	* - There are network/firewall issues preventing access
	*
	* @extends CopilotKitError
	*/
	var CopilotKitApiDiscoveryError = class extends CopilotKitError {
		constructor(params = {}) {
			var _params$url, _params$message, _params$code;
			const url = (_params$url = params.url) !== null && _params$url !== void 0 ? _params$url : "";
			let operationSuffix = "";
			if (url === null || url === void 0 ? void 0 : url.includes("/info")) operationSuffix = `when fetching CopilotKit info`;
			else if (url.includes("/actions/execute")) operationSuffix = `when attempting to execute actions.`;
			else if (url.includes("/agents/state")) operationSuffix = `when attempting to get agent state.`;
			else if (url.includes("/agents/execute")) operationSuffix = `when attempting to execute agent(s).`;
			const message = (_params$message = params.message) !== null && _params$message !== void 0 ? _params$message : params.url ? `Failed to find CopilotKit API endpoint at url ${params.url} ${operationSuffix}` : `Failed to find CopilotKit API endpoint.`;
			const code = (_params$code = params.code) !== null && _params$code !== void 0 ? _params$code : CopilotKitErrorCode.API_NOT_FOUND;
			const errorMessage = `${message}.\n\n${getSeeMoreMarkdown(ERROR_CONFIG[code].troubleshootingUrl)}`;
			super({
				message: errorMessage,
				code
			});
			this.name = ERROR_NAMES.COPILOT_API_DISCOVERY_ERROR;
		}
	};
	/**
	* This error is used for endpoints specified in runtime's remote endpoints. If they cannot be contacted
	* This typically occurs when:
	* - The API endpoint URL is invalid or misconfigured
	* - The API service is not running at the expected location
	*
	* @extends CopilotKitApiDiscoveryError
	*/
	var CopilotKitRemoteEndpointDiscoveryError = class extends CopilotKitApiDiscoveryError {
		constructor(params) {
			var _params$message2;
			const message = (_params$message2 = params === null || params === void 0 ? void 0 : params.message) !== null && _params$message2 !== void 0 ? _params$message2 : (params === null || params === void 0 ? void 0 : params.url) ? `Failed to find or contact remote endpoint at url ${params.url}` : "Failed to find or contact remote endpoint";
			const code = CopilotKitErrorCode.REMOTE_ENDPOINT_NOT_FOUND;
			super({
				message,
				code
			});
			this.name = ERROR_NAMES.COPILOT_REMOTE_ENDPOINT_DISCOVERY_ERROR;
		}
	};
	/**
	* Error thrown when a LangGraph agent cannot be found or accessed.
	* This typically occurs when:
	* - The specified agent name does not exist in the deployment
	* - The agent configuration is invalid or missing
	* - The agent service is not properly deployed or initialized
	*
	* @extends CopilotKitError
	*/
	var CopilotKitAgentDiscoveryError = class extends CopilotKitError {
		constructor(params) {
			const { agentName, availableAgents } = params;
			const code = CopilotKitErrorCode.AGENT_NOT_FOUND;
			const seeMore = getSeeMoreMarkdown(ERROR_CONFIG[code].troubleshootingUrl);
			let message;
			if (availableAgents.length) {
				const agentList = availableAgents.map((agent) => agent.name).join(", ");
				if (agentName) message = `Agent '${agentName}' was not found. Available agents are: ${agentList}. Please verify the agent name in your configuration and ensure it matches one of the available agents.\n\n${seeMore}`;
				else message = `The requested agent was not found. Available agents are: ${agentList}. Please verify the agent name in your configuration and ensure it matches one of the available agents.\n\n${seeMore}`;
			} else message = `${agentName ? `Agent '${agentName}'` : "The requested agent"} was not found. Please set up at least one agent before proceeding. ${seeMore}`;
			super({
				message,
				code
			});
			this.name = ERROR_NAMES.COPILOT_KIT_AGENT_DISCOVERY_ERROR;
		}
	};
	/**
	* Handles low-level networking errors that occur before a request reaches the server.
	* These errors arise from issues in the underlying communication infrastructure rather than
	* application-level logic or server responses. Typically used to handle "fetch failed" errors
	* where no HTTP status code is available.
	*
	* Common scenarios include:
	* - Connection failures (ECONNREFUSED) when server is down/unreachable
	* - DNS resolution failures (ENOTFOUND) when domain can't be resolved
	* - Timeouts (ETIMEDOUT) when request takes too long
	* - Protocol/transport layer errors like SSL/TLS issues
	*/
	var CopilotKitLowLevelError = class extends CopilotKitError {
		constructor({ error, url, message }) {
			let code = CopilotKitErrorCode.NETWORK_ERROR;
			const errorCode = error.code;
			const errorMessage = message !== null && message !== void 0 ? message : resolveLowLevelErrorMessage({
				errorCode,
				url
			});
			super({
				message: errorMessage,
				code
			});
			this.name = ERROR_NAMES.COPILOT_KIT_LOW_LEVEL_ERROR;
		}
	};
	/**
	* Generic catch-all error handler for HTTP responses from the CopilotKit API where a status code is available.
	* Used when we receive an HTTP error status and wish to handle broad range of them
	*
	* This differs from CopilotKitLowLevelError in that:
	* - ResolvedCopilotKitError: Server was reached and returned an HTTP status
	* - CopilotKitLowLevelError: Error occurred before reaching server (e.g. network failure)
	*
	* @param status - The HTTP status code received from the API response
	* @param message - Optional error message to include
	* @param code - Optional specific CopilotKitErrorCode to override default behavior
	*
	* Default behavior:
	* - 400 Bad Request: Maps to CopilotKitApiDiscoveryError
	* - All other status codes: Maps to UNKNOWN error code if no specific code provided
	*/
	var ResolvedCopilotKitError = class extends CopilotKitError {
		constructor({ status, message, code, isRemoteEndpoint, url }) {
			let resolvedCode = code;
			if (!resolvedCode) switch (status) {
				case 400: throw new CopilotKitApiDiscoveryError({
					message,
					url
				});
				case 404: throw isRemoteEndpoint ? new CopilotKitRemoteEndpointDiscoveryError({
					message,
					url
				}) : new CopilotKitApiDiscoveryError({
					message,
					url
				});
				default:
					resolvedCode = CopilotKitErrorCode.UNKNOWN;
					break;
			}
			super({
				message,
				code: resolvedCode
			});
			this.name = ERROR_NAMES.RESOLVED_COPILOT_KIT_ERROR;
		}
	};
	var ConfigurationError = class extends CopilotKitError {
		constructor(message) {
			super({
				message,
				code: CopilotKitErrorCode.CONFIGURATION_ERROR
			});
			this.name = ERROR_NAMES.CONFIGURATION_ERROR;
			this.severity = Severity.WARNING;
		}
	};
	var MissingPublicApiKeyError = class extends ConfigurationError {
		constructor(message) {
			super(message);
			this.name = ERROR_NAMES.MISSING_PUBLIC_API_KEY_ERROR;
			this.severity = Severity.CRITICAL;
		}
	};
	var UpgradeRequiredError = class extends ConfigurationError {
		constructor(message) {
			super(message);
			this.name = ERROR_NAMES.UPGRADE_REQUIRED_ERROR;
			this.severity = Severity.WARNING;
		}
	};
	/**
	* Checks if an error is already a structured CopilotKit error.
	* This utility centralizes the logic for detecting structured errors across the codebase.
	*
	* @param error - The error to check
	* @returns true if the error is already structured, false otherwise
	*/
	function isStructuredCopilotKitError(error) {
		var _error$extensions;
		return error instanceof CopilotKitError || error instanceof CopilotKitLowLevelError || (error === null || error === void 0 ? void 0 : error.name) && error.name.includes("CopilotKit") || (error === null || error === void 0 || (_error$extensions = error.extensions) === null || _error$extensions === void 0 ? void 0 : _error$extensions.code) !== void 0;
	}
	/**
	* Returns the error as-is if it's already structured, otherwise converts it using the provided converter function.
	* This utility centralizes the pattern of preserving structured errors while converting unstructured ones.
	*
	* @param error - The error to process
	* @param converter - Function to convert unstructured errors to structured ones
	* @returns The structured error
	*/
	function ensureStructuredError(error, converter) {
		return isStructuredCopilotKitError(error) ? error : converter(error);
	}
	async function getPossibleVersionMismatch({ runtimeVersion, runtimeClientGqlVersion }) {
		if (!runtimeVersion || runtimeVersion === "" || !runtimeClientGqlVersion) return;
		if (COPILOTKIT_VERSION !== runtimeVersion || COPILOTKIT_VERSION !== runtimeClientGqlVersion || runtimeVersion !== runtimeClientGqlVersion) return {
			runtimeVersion,
			runtimeClientGqlVersion,
			reactCoreVersion: COPILOTKIT_VERSION,
			message: getVersionMismatchErrorMessage({
				runtimeVersion,
				runtimeClientGqlVersion,
				reactCoreVersion: COPILOTKIT_VERSION
			})
		};
	}
	const resolveLowLevelErrorMessage = ({ errorCode, url }) => {
		const troubleshootingLink = ERROR_CONFIG[CopilotKitErrorCode.NETWORK_ERROR].troubleshootingUrl;
		const genericMessage = (description = `Failed to fetch from url ${url}.`) => `${description}.

Possible reasons:
- -The server may have an error preventing it from returning a response (Check the server logs for more info).
- -The server might be down or unreachable
- -There might be a network issue (e.g., DNS failure, connection timeout) 
- -The URL might be incorrect
- -The server is not running on the specified port

${getSeeMoreMarkdown(troubleshootingLink)}`;
		if (url.includes("/info")) return genericMessage(`Failed to fetch CopilotKit agents/action information from url ${url}.`);
		if (url.includes("/actions/execute")) return genericMessage(`Fetch call to ${url} to execute actions failed.`);
		if (url.includes("/agents/state")) return genericMessage(`Fetch call to ${url} to get agent state failed.`);
		if (url.includes("/agents/execute")) return genericMessage(`Fetch call to ${url} to execute agent(s) failed.`);
		switch (errorCode) {
			case "ECONNREFUSED": return `Connection to ${url} was refused. Ensure the server is running and accessible.\n\n${getSeeMoreMarkdown(troubleshootingLink)}`;
			case "ENOTFOUND": return `The server on ${url} could not be found. Check the URL or your network configuration.\n\n${getSeeMoreMarkdown(ERROR_CONFIG[CopilotKitErrorCode.NOT_FOUND].troubleshootingUrl)}`;
			case "ETIMEDOUT": return `The connection to ${url} timed out. The server might be overloaded or taking too long to respond.\n\n${getSeeMoreMarkdown(troubleshootingLink)}`;
			default: return;
		}
	};

//#endregion
//#region src/utils/json-schema.ts
	function actionParametersToJsonSchema(actionParameters) {
		let parameters = {};
		for (let parameter of actionParameters || []) parameters[parameter.name] = convertAttribute(parameter);
		let requiredParameterNames = [];
		for (let arg of actionParameters || []) if (arg.required !== false) requiredParameterNames.push(arg.name);
		return {
			type: "object",
			properties: parameters,
			required: requiredParameterNames
		};
	}
	function jsonSchemaToActionParameters(jsonSchema) {
		if (jsonSchema.type !== "object" || !jsonSchema.properties) return [];
		const parameters = [];
		const requiredFields = jsonSchema.required || [];
		for (const [name, schema] of Object.entries(jsonSchema.properties)) {
			const parameter = convertJsonSchemaToParameter(name, schema, requiredFields.includes(name));
			parameters.push(parameter);
		}
		return parameters;
	}
	function convertJsonSchemaToParameter(name, schema, isRequired) {
		const baseParameter = {
			name,
			description: schema.description
		};
		if (!isRequired) baseParameter.required = false;
		if (Array.isArray(schema.type)) {
			const types = schema.type;
			const hasNull = types.includes("null");
			const nonNullTypes = types.filter((t) => t !== "null");
			const resolvedType = nonNullTypes.length > 0 ? nonNullTypes[0] : "string";
			return convertJsonSchemaToParameter(name, {
				...schema,
				type: resolvedType
			}, hasNull ? false : isRequired);
		}
		switch (schema.type) {
			case "string": return {
				...baseParameter,
				type: "string",
				...schema.enum && { enum: schema.enum }
			};
			case "number":
			case "boolean": return {
				...baseParameter,
				type: schema.type
			};
			case "object":
				if (schema.properties) {
					const attributes = [];
					const requiredFields = schema.required || [];
					for (const [propName, propSchema] of Object.entries(schema.properties)) attributes.push(convertJsonSchemaToParameter(propName, propSchema, requiredFields.includes(propName)));
					return {
						...baseParameter,
						type: "object",
						attributes
					};
				}
				return {
					...baseParameter,
					type: "object"
				};
			case "array": if (schema.items.type === "object" && "properties" in schema.items) {
				const attributes = [];
				const requiredFields = schema.items.required || [];
				for (const [propName, propSchema] of Object.entries(schema.items.properties || {})) attributes.push(convertJsonSchemaToParameter(propName, propSchema, requiredFields.includes(propName)));
				return {
					...baseParameter,
					type: "object[]",
					attributes
				};
			} else if (schema.items.type === "array") throw new Error("Nested arrays are not supported");
			else return {
				...baseParameter,
				type: `${schema.items.type}[]`
			};
			default: return {
				...baseParameter,
				type: "string"
			};
		}
	}
	function convertAttribute(attribute) {
		switch (attribute.type) {
			case "string": return {
				type: "string",
				description: attribute.description,
				...attribute.enum && { enum: attribute.enum }
			};
			case "number":
			case "boolean": return {
				type: attribute.type,
				description: attribute.description
			};
			case "object":
			case "object[]":
				var _attribute$attributes, _attribute$attributes2;
				const properties = (_attribute$attributes = attribute.attributes) === null || _attribute$attributes === void 0 ? void 0 : _attribute$attributes.reduce((acc, attr) => {
					acc[attr.name] = convertAttribute(attr);
					return acc;
				}, {});
				const required = (_attribute$attributes2 = attribute.attributes) === null || _attribute$attributes2 === void 0 ? void 0 : _attribute$attributes2.filter((attr) => attr.required !== false).map((attr) => attr.name);
				if (attribute.type === "object[]") return {
					type: "array",
					items: {
						type: "object",
						...properties && { properties },
						...required && required.length > 0 && { required }
					},
					description: attribute.description
				};
				return {
					type: "object",
					description: attribute.description,
					...properties && { properties },
					...required && required.length > 0 && { required }
				};
			default:
				var _attribute$type;
				if ((_attribute$type = attribute.type) === null || _attribute$type === void 0 ? void 0 : _attribute$type.endsWith("[]")) return {
					type: "array",
					items: { type: attribute.type.slice(0, -2) },
					description: attribute.description
				};
				return {
					type: "string",
					description: attribute.description
				};
		}
	}
	function convertJsonSchemaToZodSchema(jsonSchema, required, definitions, visitedRefs) {
		var _ref, _jsonSchema$anyOf, _jsonSchema$type;
		if (jsonSchema.$ref && definitions) {
			const refPath = jsonSchema.$ref.replace(/^#\/\$defs\/|^#\/definitions\//, "");
			const refs = visitedRefs !== null && visitedRefs !== void 0 ? visitedRefs : /* @__PURE__ */ new Set();
			if (refs.has(refPath)) {
				console.warn(`[CopilotKit] Circular $ref detected for "${refPath}" — falling back to z.any()`);
				let schema = zod.z.any();
				if (jsonSchema.description) schema = schema.describe(jsonSchema.description);
				return required ? schema : schema.optional();
			}
			const resolved = definitions[refPath];
			if (resolved) {
				const nextRefs = new Set(refs);
				nextRefs.add(refPath);
				return convertJsonSchemaToZodSchema(resolved, required, definitions, nextRefs);
			}
		}
		const defs = (_ref = definitions !== null && definitions !== void 0 ? definitions : jsonSchema.$defs) !== null && _ref !== void 0 ? _ref : jsonSchema.definitions;
		if (Array.isArray(jsonSchema.type)) {
			const types = jsonSchema.type;
			const hasNull = types.includes("null");
			const nonNullTypes = types.filter((t) => t !== "null");
			const resolvedType = nonNullTypes.length > 0 ? nonNullTypes[0] : "string";
			const innerSchema = convertJsonSchemaToZodSchema({
				...jsonSchema,
				type: resolvedType
			}, true, defs, visitedRefs);
			let schema = hasNull ? zod.z.union([innerSchema, zod.z.null()]) : innerSchema;
			if (jsonSchema.description) schema = schema.describe(jsonSchema.description);
			return required ? schema : schema.optional();
		}
		const unionVariants = (_jsonSchema$anyOf = jsonSchema.anyOf) !== null && _jsonSchema$anyOf !== void 0 ? _jsonSchema$anyOf : jsonSchema.oneOf;
		if (Array.isArray(unionVariants) && unionVariants.length > 0) {
			if (unionVariants.length === 1) return convertJsonSchemaToZodSchema(unionVariants[0], required, defs, visitedRefs);
			const schemas = unionVariants.map((v) => convertJsonSchemaToZodSchema(v, true, defs, visitedRefs));
			let schema = zod.z.union(schemas);
			if (jsonSchema.description) schema = schema.describe(jsonSchema.description);
			return required ? schema : schema.optional();
		}
		if (jsonSchema.type === "object") {
			const spec = {};
			if (!jsonSchema.properties || !Object.keys(jsonSchema.properties).length) return !required ? zod.z.object(spec).optional() : zod.z.object(spec);
			for (const [key, value] of Object.entries(jsonSchema.properties)) spec[key] = convertJsonSchemaToZodSchema(value, jsonSchema.required ? jsonSchema.required.includes(key) : false, defs, visitedRefs);
			let schema = zod.z.object(spec).describe(jsonSchema.description);
			return required ? schema : schema.optional();
		} else if (jsonSchema.type === "string") {
			if (jsonSchema.enum && jsonSchema.enum.length > 0) {
				let schema = zod.z.enum(jsonSchema.enum).describe(jsonSchema.description);
				return required ? schema : schema.optional();
			}
			let schema = zod.z.string().describe(jsonSchema.description);
			return required ? schema : schema.optional();
		} else if (jsonSchema.type === "number" || jsonSchema.type === "integer") {
			let schema = zod.z.number().describe(jsonSchema.description);
			return required ? schema : schema.optional();
		} else if (jsonSchema.type === "boolean") {
			let schema = zod.z.boolean().describe(jsonSchema.description);
			return required ? schema : schema.optional();
		} else if (jsonSchema.type === "array") {
			let itemSchema = convertJsonSchemaToZodSchema(jsonSchema.items, true, defs, visitedRefs);
			let schema = zod.z.array(itemSchema).describe(jsonSchema.description);
			return required ? schema : schema.optional();
		} else if (jsonSchema.type === "null") {
			let schema = zod.z.null().describe(jsonSchema.description);
			return required ? schema : schema.optional();
		}
		console.warn(`[CopilotKit] Unsupported JSON schema type "${(_jsonSchema$type = jsonSchema.type) !== null && _jsonSchema$type !== void 0 ? _jsonSchema$type : "unknown"}" — falling back to z.any()`);
		let schema = zod.z.any();
		if (jsonSchema.description) schema = schema.describe(jsonSchema.description);
		return required ? schema : schema.optional();
	}
	function getZodParameters(parameters) {
		if (!parameters) return zod.z.object({});
		return convertJsonSchemaToZodSchema(actionParametersToJsonSchema(parameters), true);
	}

//#endregion
//#region src/utils/types.ts
	const RUNTIME_MODE_SSE = "sse";
	const RUNTIME_MODE_INTELLIGENCE = "intelligence";

//#endregion
//#region src/utils/random-id.ts
	function randomId() {
		return "ck-" + (0, uuid.v4)();
	}
	function randomUUID() {
		return (0, uuid.v4)();
	}
	/**
	* Recursively converts an object to a serializable form by converting functions to their string representation.
	*/
	function toSerializable(value) {
		if (typeof value === "function") return value.toString();
		if (Array.isArray(value)) return value.map(toSerializable);
		if (value !== null && typeof value === "object") {
			const result = {};
			for (const key of Object.keys(value)) result[key] = toSerializable(value[key]);
			return result;
		}
		return value;
	}
	function dataToUUID(input, namespace) {
		const BASE_NAMESPACE = "e4b01160-ff74-4c6e-9b27-d53cd930fe8e";
		const boundNamespace = namespace ? (0, uuid.v5)(namespace, BASE_NAMESPACE) : BASE_NAMESPACE;
		return (0, uuid.v5)(typeof input === "string" ? input : JSON.stringify(toSerializable(input)), boundNamespace);
	}
	function isValidUUID(uuid$1) {
		return (0, uuid.validate)(uuid$1);
	}

//#endregion
//#region src/utils/requests.ts
/**
	* Safely read a Response/Request body with sensible defaults:
	* - clones the response/request to avoid consuming the original response/request
	* - Skips GET/HEAD
	* - Tries JSON first regardless of content-type
	* - Falls back to text and optionally parses when it "looks" like JSON
	*/
	async function readBody(r) {
		const method = "method" in r ? r.method.toUpperCase() : void 0;
		if (method === "GET" || method === "HEAD") return;
		if (!("body" in r) || r.body == null) return;
		try {
			return await r.clone().json();
		} catch (_unused) {
			try {
				const text = await r.clone().text();
				const trimmed = text.trim();
				if (trimmed.length === 0) return text;
				if (trimmed.startsWith("{") || trimmed.startsWith("[")) try {
					return JSON.parse(trimmed);
				} catch (_unused2) {
					return text;
				}
				return text;
			} catch (_unused3) {
				try {
					var _c$body;
					const stream = (_c$body = r.clone().body) !== null && _c$body !== void 0 ? _c$body : null;
					if (!stream) return void 0;
					const reader = stream.getReader();
					const decoder = new TextDecoder();
					let out = "";
					while (true) {
						const { done, value } = await reader.read();
						if (done) break;
						if (typeof value === "string") out += value;
						else out += decoder.decode(value, { stream: true });
					}
					out += decoder.decode();
					const trimmed = out.trim();
					if (trimmed.length === 0) return out;
					if (trimmed.startsWith("{") || trimmed.startsWith("[")) try {
						return JSON.parse(trimmed);
					} catch (_unused4) {
						return out;
					}
					return out;
				} catch (_unused5) {
					return;
				}
			}
		}
	}

//#endregion
//#region src/utils/index.ts
/**
	* Safely parses a JSON string into an object
	* @param json The JSON string to parse
	* @param fallback Optional fallback value to return if parsing fails. If not provided or set to "unset", returns null
	* @returns The parsed JSON object, or the fallback value (or null) if parsing fails
	*/
	function parseJson(json, fallback = "unset") {
		try {
			return JSON.parse(json);
		} catch (e) {
			return fallback === "unset" ? null : fallback;
		}
	}
	/**
	* Parses a partial/incomplete JSON string, returning as much valid data as possible.
	* Falls back to an empty object if parsing fails entirely.
	*/
	function partialJSONParse(json) {
		try {
			const parsed = partial_json.parse(json);
			if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) return parsed;
			return {};
		} catch (error) {
			return {};
		}
	}
	/**
	* Returns an exponential backoff function suitable for Phoenix.js
	* `reconnectAfterMs` and `rejoinAfterMs` options.
	*
	* @param baseMs  - Initial delay for the first retry attempt.
	* @param maxMs   - Upper bound — delays are capped at this value.
	*
	* Phoenix calls the returned function with a 1-based `tries` count.
	* The delay doubles on each attempt: baseMs, 2×baseMs, 4×baseMs, …, maxMs.
	*/
	function phoenixExponentialBackoff(baseMs, maxMs) {
		return (tries) => Math.min(baseMs * 2 ** (tries - 1), maxMs);
	}
	/**
	* Maps an array of items to a new array, skipping items that throw errors during mapping
	* @param items The array to map
	* @param callback The mapping function to apply to each item
	* @returns A new array containing only the successfully mapped items
	*/
	function tryMap(items, callback) {
		return items.reduce((acc, item, index, array) => {
			try {
				acc.push(callback(item, index, array));
			} catch (error) {
				console.error(error);
			}
			return acc;
		}, []);
	}
	/**
	* Checks if the current environment is macOS
	* @returns {boolean} True if running on macOS, false otherwise
	*/
	function isMacOS() {
		return /Mac|iMac|Macintosh/i.test(navigator.userAgent);
	}
	/**
	* Safely parses a JSON string into a tool arguments object.
	* Returns the parsed object only if it's a plain object (not an array, null, etc.).
	* Falls back to an empty object for any non-object JSON value or parse failure.
	*/
	function safeParseToolArgs(raw) {
		try {
			const parsed = JSON.parse(raw);
			if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) return parsed;
			console.warn(`[CopilotKit] Tool arguments parsed to non-object (${typeof parsed}), falling back to empty object`);
			return {};
		} catch (_unused) {
			console.warn("[CopilotKit] Failed to parse tool arguments, falling back to empty object");
			return {};
		}
	}

//#endregion
//#region src/constants/index.ts
	const COPILOT_CLOUD_API_URL = "https://api.cloud.copilotkit.ai";
	const COPILOT_CLOUD_VERSION = "v1";
	const COPILOT_CLOUD_CHAT_URL = `${COPILOT_CLOUD_API_URL}/copilotkit/${COPILOT_CLOUD_VERSION}`;
	const COPILOT_CLOUD_PUBLIC_API_KEY_HEADER = "X-CopilotCloud-Public-Api-Key";
	const DEFAULT_AGENT_ID = "default";
	/** Phoenix channel event name used for all AG-UI events. */
	const AG_UI_CHANNEL_EVENT = "ag-ui";

//#endregion
//#region src/telemetry/utils.ts
	function flattenObject(obj, parentKey = "", res = {}) {
		for (let key in obj) {
			const propName = parentKey ? `${parentKey}.${key}` : key;
			if (typeof obj[key] === "object" && obj[key] !== null) flattenObject(obj[key], propName, res);
			else res[propName] = obj[key];
		}
		return res;
	}

//#endregion
//#region package.json
	var version = "1.57.1";

//#endregion
//#region src/telemetry/scarf-client.ts
	const SCARF_BASE_URL = `https://copilotkit.gateway.scarf.sh/${version}`;
	var ScarfClient = class {
		constructor() {}
		async logEvent(properties) {
			try {
				const controller = new AbortController();
				const timeoutId = setTimeout(() => controller.abort(), 3e3);
				const queryParams = new URLSearchParams();
				Object.entries(properties).forEach(([key, value]) => {
					if (value !== null && value !== void 0) queryParams.append(key, String(value));
				});
				const url = `${SCARF_BASE_URL}?${queryParams.toString()}`;
				const response = await fetch(url, {
					method: "GET",
					signal: controller.signal
				});
				clearTimeout(timeoutId);
				if (!response.ok) throw new Error(`HTTP error! status: ${response.status}`);
			} catch (_unused) {}
		}
	};
	var scarf_client_default = new ScarfClient();

//#endregion
//#region src/telemetry/telemetry-client.ts
/**
	* Checks if telemetry is disabled via environment variables.
	* Users can opt out by setting:
	* - COPILOTKIT_TELEMETRY_DISABLED=true or COPILOTKIT_TELEMETRY_DISABLED=1
	* - DO_NOT_TRACK=true or DO_NOT_TRACK=1
	*/
	function isTelemetryDisabled() {
		return process.env.COPILOTKIT_TELEMETRY_DISABLED === "true" || process.env.COPILOTKIT_TELEMETRY_DISABLED === "1" || process.env.DO_NOT_TRACK === "true" || process.env.DO_NOT_TRACK === "1";
	}
	var TelemetryClient = class {
		constructor({ packageName, packageVersion, telemetryDisabled, telemetryBaseUrl, sampleRate }) {
			this.globalProperties = {};
			this.cloudConfiguration = null;
			this.telemetryDisabled = false;
			this.sampleRate = .05;
			this.anonymousId = `anon_${(0, uuid.v4)()}`;
			this.packageName = packageName;
			this.packageVersion = packageVersion;
			this.telemetryDisabled = telemetryDisabled || isTelemetryDisabled();
			if (this.telemetryDisabled) return;
			this.setSampleRate(sampleRate);
			this.segment = new _segment_analytics_node.Analytics({ writeKey: process.env.COPILOTKIT_SEGMENT_WRITE_KEY || "n7XAZtQCGS2v1vvBy3LgBCv2h3Y8whja" });
			this.setGlobalProperties({
				"copilotkit.package.name": packageName,
				"copilotkit.package.version": packageVersion
			});
		}
		shouldSendEvent() {
			return Math.random() < this.sampleRate;
		}
		async capture(event, properties) {
			if (!this.shouldSendEvent() || !this.segment) return;
			const flattenedProperties = flattenObject(properties);
			const propertiesWithGlobal = {
				...this.globalProperties,
				...flattenedProperties
			};
			const orderedPropertiesWithGlobal = Object.keys(propertiesWithGlobal).sort().reduce((obj, key) => {
				obj[key] = propertiesWithGlobal[key];
				return obj;
			}, {});
			this.segment.track({
				anonymousId: this.anonymousId,
				event,
				properties: { ...orderedPropertiesWithGlobal }
			});
			await scarf_client_default.logEvent({ event });
		}
		setGlobalProperties(properties) {
			const flattenedProperties = flattenObject(properties);
			this.globalProperties = {
				...this.globalProperties,
				...flattenedProperties
			};
		}
		setCloudConfiguration(properties) {
			this.cloudConfiguration = properties;
			this.setGlobalProperties({ cloud: {
				publicApiKey: properties.publicApiKey,
				baseUrl: properties.baseUrl
			} });
		}
		setSampleRate(sampleRate) {
			let _sampleRate;
			_sampleRate = sampleRate !== null && sampleRate !== void 0 ? sampleRate : .05;
			if (process.env.COPILOTKIT_TELEMETRY_SAMPLE_RATE) _sampleRate = parseFloat(process.env.COPILOTKIT_TELEMETRY_SAMPLE_RATE);
			if (_sampleRate < 0 || _sampleRate > 1) throw new Error("Sample rate must be between 0 and 1");
			this.sampleRate = _sampleRate;
			this.setGlobalProperties({
				sampleRate: this.sampleRate,
				sampleRateAdjustmentFactor: 1 - this.sampleRate,
				sampleWeight: 1 / this.sampleRate
			});
		}
	};

//#endregion
//#region src/debug.ts
/** The all-off config used when debug is falsy. */
	const DEBUG_OFF = {
		enabled: false,
		events: false,
		lifecycle: false,
		verbose: false
	};
	/**
	* Normalizes a DebugConfig value into a ResolvedDebugConfig.
	*
	* - `false` / `undefined` → all off
	* - `true` → events + lifecycle on, verbose off (no PII in logs)
	* - object → merges with defaults (events: true, lifecycle: true, verbose: false)
	*/
	function resolveDebugConfig(debug) {
		var _debug$events, _debug$lifecycle, _debug$verbose;
		if (!debug) return DEBUG_OFF;
		if (debug === true) return {
			enabled: true,
			events: true,
			lifecycle: true,
			verbose: false
		};
		const events = (_debug$events = debug.events) !== null && _debug$events !== void 0 ? _debug$events : true;
		const lifecycle = (_debug$lifecycle = debug.lifecycle) !== null && _debug$lifecycle !== void 0 ? _debug$lifecycle : true;
		const enabled = events || lifecycle;
		return {
			enabled,
			events,
			lifecycle,
			verbose: enabled && ((_debug$verbose = debug.verbose) !== null && _debug$verbose !== void 0 ? _debug$verbose : false)
		};
	}

//#endregion
//#region src/standard-schema.ts
/**
	* Check whether a schema implements the Standard JSON Schema V1 protocol.
	*/
	function hasStandardJsonSchema(schema) {
		const props = schema["~standard"];
		return props != null && typeof props === "object" && "jsonSchema" in props && props.jsonSchema != null && typeof props.jsonSchema === "object" && "input" in props.jsonSchema && typeof props.jsonSchema.input === "function";
	}
	/**
	* Convert any StandardSchemaV1-compatible schema to a JSON Schema object.
	*
	* Strategy:
	* 1. If the schema implements Standard JSON Schema V1 (`~standard.jsonSchema`),
	*    call `schema['~standard'].jsonSchema.input({ target: 'draft-07' })`.
	* 2. If the schema exposes a `toJSONSchema()` method (Zod v4), call it directly.
	* 3. If the schema is a Zod v3 schema (`~standard.vendor === 'zod'`), use the
	*    injected `zodToJsonSchema()` function.
	* 4. Otherwise throw a descriptive error.
	*/
	function schemaToJsonSchema(schema, options) {
		if (hasStandardJsonSchema(schema)) return schema["~standard"].jsonSchema.input({ target: "draft-07" });
		if (typeof schema.toJSONSchema === "function") return schema.toJSONSchema();
		const vendor = schema["~standard"].vendor;
		if (vendor === "zod" && (options === null || options === void 0 ? void 0 : options.zodToJsonSchema)) return options.zodToJsonSchema(schema, { $refStrategy: "none" });
		throw new Error(`Cannot convert schema to JSON Schema. The schema (vendor: "${vendor}") does not implement Standard JSON Schema V1 and no zodToJsonSchema fallback is available. Use a library that supports Standard JSON Schema (e.g., Zod 3.24+, Valibot v1+, ArkType v2+) or pass a zodToJsonSchema function in options.`);
	}

//#endregion
//#region src/attachments/utils.ts
	const DEFAULT_MAX_SIZE = 20 * 1024 * 1024;
	/**
	* Derive the attachment modality from a MIME type string.
	*/
	function getModalityFromMimeType(mimeType) {
		if (mimeType.startsWith("image/")) return "image";
		if (mimeType.startsWith("audio/")) return "audio";
		if (mimeType.startsWith("video/")) return "video";
		return "document";
	}
	/**
	* Format a byte count as a human-readable file size string.
	*/
	function formatFileSize(bytes) {
		if (bytes < 1024) return `${bytes} B`;
		if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
		return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
	}
	/**
	* Check if a file exceeds the maximum allowed size.
	*/
	function exceedsMaxSize(file, maxSize = DEFAULT_MAX_SIZE) {
		return file.size > maxSize;
	}
	/**
	* Read a File as a base64 string (without the data URL prefix).
	*/
	function readFileAsBase64(file) {
		return new Promise((resolve, reject) => {
			const reader = new FileReader();
			reader.onload = (e) => {
				var _e$target;
				const result = (_e$target = e.target) === null || _e$target === void 0 ? void 0 : _e$target.result;
				const base64 = result === null || result === void 0 ? void 0 : result.split(",")[1];
				if (base64) resolve(base64);
				else reject(/* @__PURE__ */ new Error("Failed to read file as base64"));
			};
			reader.onerror = reject;
			reader.readAsDataURL(file);
		});
	}
	/**
	* Generate a thumbnail data URL from a video file by capturing a frame near the start (at 0.1s).
	* Returns undefined if thumbnail generation fails or if called outside a browser environment.
	*/
	function generateVideoThumbnail(file) {
		if (typeof document === "undefined") return Promise.resolve(void 0);
		return new Promise((resolve) => {
			let resolved = false;
			const video = document.createElement("video");
			const canvas = document.createElement("canvas");
			const url = URL.createObjectURL(file);
			const cleanup = (result) => {
				if (resolved) return;
				resolved = true;
				URL.revokeObjectURL(url);
				resolve(result);
			};
			const timeout = setTimeout(() => {
				console.warn(`[CopilotKit] generateVideoThumbnail: timed out for file "${file.name}"`);
				cleanup(void 0);
			}, 1e4);
			video.preload = "metadata";
			video.muted = true;
			video.playsInline = true;
			video.onloadeddata = () => {
				video.currentTime = .1;
			};
			video.onseeked = () => {
				clearTimeout(timeout);
				canvas.width = video.videoWidth;
				canvas.height = video.videoHeight;
				const ctx = canvas.getContext("2d");
				if (ctx) {
					ctx.drawImage(video, 0, 0);
					cleanup(canvas.toDataURL("image/jpeg", .7));
				} else {
					console.warn("[CopilotKit] generateVideoThumbnail: could not get 2d canvas context");
					cleanup(void 0);
				}
			};
			video.onerror = () => {
				clearTimeout(timeout);
				console.warn(`[CopilotKit] generateVideoThumbnail: video element error for file "${file.name}"`);
				cleanup(void 0);
			};
			video.src = url;
		});
	}
	/**
	* Check if a file's MIME type matches an accept filter string.
	* Handles file extensions (e.g. ".pdf"), MIME wildcards ("image/*"), and comma-separated lists.
	*/
	function matchesAcceptFilter(file, accept) {
		if (!accept || accept === "*/*") return true;
		return accept.split(",").map((f) => f.trim()).some((filter) => {
			if (filter.startsWith(".")) {
				var _file$name;
				return ((_file$name = file.name) !== null && _file$name !== void 0 ? _file$name : "").toLowerCase().endsWith(filter.toLowerCase());
			}
			if (filter.endsWith("/*")) {
				const prefix = filter.slice(0, -2);
				return file.type.startsWith(prefix + "/");
			}
			return file.type === filter;
		});
	}
	/**
	* Convert an InputContentSource to a usable URL string.
	* For data sources, returns a base64 data URL; for URL sources, returns the URL directly.
	*/
	function getSourceUrl(source) {
		if (source.type === "url") return source.value;
		return `data:${source.mimeType};base64,${source.value}`;
	}
	/**
	* Return a short human-readable label for a document MIME type (e.g. "PDF", "DOC").
	*/
	function getDocumentIcon(mimeType) {
		if (mimeType.includes("pdf")) return "PDF";
		if (mimeType.includes("sheet") || mimeType.includes("excel")) return "XLS";
		if (mimeType.includes("presentation") || mimeType.includes("powerpoint")) return "PPT";
		if (mimeType.includes("word") || mimeType.includes("document")) return "DOC";
		if (mimeType.includes("text/")) return "TXT";
		return "FILE";
	}

//#endregion
//#region src/logger.ts
	const logger = console;

//#endregion
//#region src/finalize-events.ts
	const defaultStopMessage = "Run stopped by user";
	const defaultAbruptEndMessage = "Run ended without emitting a terminal event";
	function finalizeRunEvents(events, options = {}) {
		const { stopRequested = false, interruptionMessage } = options;
		const resolvedStopMessage = interruptionMessage !== null && interruptionMessage !== void 0 ? interruptionMessage : defaultStopMessage;
		const resolvedAbruptMessage = interruptionMessage && interruptionMessage !== defaultStopMessage ? interruptionMessage : defaultAbruptEndMessage;
		const appended = [];
		const openMessageIds = /* @__PURE__ */ new Set();
		const openToolCalls = /* @__PURE__ */ new Map();
		for (const event of events) switch (event.type) {
			case _ag_ui_client.EventType.TEXT_MESSAGE_START: {
				const messageId = event.messageId;
				if (typeof messageId === "string") openMessageIds.add(messageId);
				break;
			}
			case _ag_ui_client.EventType.TEXT_MESSAGE_END: {
				const messageId = event.messageId;
				if (typeof messageId === "string") openMessageIds.delete(messageId);
				break;
			}
			case _ag_ui_client.EventType.TOOL_CALL_START: {
				const toolCallId = event.toolCallId;
				if (typeof toolCallId === "string") openToolCalls.set(toolCallId, {
					hasEnd: false,
					hasResult: false
				});
				break;
			}
			case _ag_ui_client.EventType.TOOL_CALL_END: {
				const toolCallId = event.toolCallId;
				const info = toolCallId ? openToolCalls.get(toolCallId) : void 0;
				if (info) info.hasEnd = true;
				break;
			}
			case _ag_ui_client.EventType.TOOL_CALL_RESULT: {
				const toolCallId = event.toolCallId;
				const info = toolCallId ? openToolCalls.get(toolCallId) : void 0;
				if (info) info.hasResult = true;
				break;
			}
			default: break;
		}
		const hasRunFinished = events.some((event) => event.type === _ag_ui_client.EventType.RUN_FINISHED);
		const hasRunError = events.some((event) => event.type === _ag_ui_client.EventType.RUN_ERROR);
		const terminalEventMissing = !(hasRunFinished || hasRunError);
		for (const messageId of openMessageIds) {
			const endEvent = {
				type: _ag_ui_client.EventType.TEXT_MESSAGE_END,
				messageId
			};
			events.push(endEvent);
			appended.push(endEvent);
		}
		for (const [toolCallId, info] of openToolCalls) {
			if (!info.hasEnd) {
				const endEvent = {
					type: _ag_ui_client.EventType.TOOL_CALL_END,
					toolCallId
				};
				events.push(endEvent);
				appended.push(endEvent);
			}
			if (terminalEventMissing && !info.hasResult) {
				const resultEvent = {
					type: _ag_ui_client.EventType.TOOL_CALL_RESULT,
					toolCallId,
					messageId: `${toolCallId !== null && toolCallId !== void 0 ? toolCallId : randomUUID()}-result`,
					role: "tool",
					content: JSON.stringify(stopRequested ? {
						status: "stopped",
						reason: "stop_requested",
						message: resolvedStopMessage
					} : {
						status: "error",
						reason: "missing_terminal_event",
						message: resolvedAbruptMessage
					})
				};
				events.push(resultEvent);
				appended.push(resultEvent);
			}
		}
		if (terminalEventMissing) if (stopRequested) {
			const finishedEvent = { type: _ag_ui_client.EventType.RUN_FINISHED };
			events.push(finishedEvent);
			appended.push(finishedEvent);
		} else {
			const errorEvent = {
				type: _ag_ui_client.EventType.RUN_ERROR,
				message: resolvedAbruptMessage,
				code: "INCOMPLETE_STREAM"
			};
			events.push(errorEvent);
			appended.push(errorEvent);
		}
		return appended;
	}

//#endregion
//#region src/transcription-errors.ts
/**
	* Error codes for transcription HTTP responses.
	* Uses snake_case to align with existing CopilotKitCoreErrorCode pattern.
	* These codes are returned by the runtime and parsed by the client.
	*/
	let TranscriptionErrorCode = /* @__PURE__ */ function(TranscriptionErrorCode) {
		/** Transcription service not configured in runtime */
		TranscriptionErrorCode["SERVICE_NOT_CONFIGURED"] = "service_not_configured";
		/** Audio format not supported */
		TranscriptionErrorCode["INVALID_AUDIO_FORMAT"] = "invalid_audio_format";
		/** Audio file is too long */
		TranscriptionErrorCode["AUDIO_TOO_LONG"] = "audio_too_long";
		/** Audio file is empty or too short */
		TranscriptionErrorCode["AUDIO_TOO_SHORT"] = "audio_too_short";
		/** Rate limited by transcription provider */
		TranscriptionErrorCode["RATE_LIMITED"] = "rate_limited";
		/** Authentication failed with transcription provider */
		TranscriptionErrorCode["AUTH_FAILED"] = "auth_failed";
		/** Transcription provider returned an error */
		TranscriptionErrorCode["PROVIDER_ERROR"] = "provider_error";
		/** Network error during transcription */
		TranscriptionErrorCode["NETWORK_ERROR"] = "network_error";
		/** Invalid request format */
		TranscriptionErrorCode["INVALID_REQUEST"] = "invalid_request";
		return TranscriptionErrorCode;
	}({});
	/**
	* Helper functions to create transcription error responses.
	* Used by the runtime to return consistent error responses.
	*/
	const TranscriptionErrors = {
		serviceNotConfigured: () => ({
			error: TranscriptionErrorCode.SERVICE_NOT_CONFIGURED,
			message: "Transcription service is not configured",
			retryable: false
		}),
		invalidAudioFormat: (format, supported) => ({
			error: TranscriptionErrorCode.INVALID_AUDIO_FORMAT,
			message: `Unsupported audio format: ${format}. Supported: ${supported.join(", ")}`,
			retryable: false
		}),
		invalidRequest: (details) => ({
			error: TranscriptionErrorCode.INVALID_REQUEST,
			message: details,
			retryable: false
		}),
		rateLimited: () => ({
			error: TranscriptionErrorCode.RATE_LIMITED,
			message: "Rate limited. Please try again later.",
			retryable: true
		}),
		authFailed: () => ({
			error: TranscriptionErrorCode.AUTH_FAILED,
			message: "Authentication failed with transcription provider",
			retryable: false
		}),
		providerError: (message) => ({
			error: TranscriptionErrorCode.PROVIDER_ERROR,
			message,
			retryable: true
		}),
		networkError: (message = "Network error during transcription") => ({
			error: TranscriptionErrorCode.NETWORK_ERROR,
			message,
			retryable: true
		}),
		audioTooLong: () => ({
			error: TranscriptionErrorCode.AUDIO_TOO_LONG,
			message: "Audio file is too long",
			retryable: false
		}),
		audioTooShort: () => ({
			error: TranscriptionErrorCode.AUDIO_TOO_SHORT,
			message: "Audio is too short to transcribe",
			retryable: false
		})
	};

//#endregion
//#region src/a2ui-prompts.ts
/**
	* Default A2UI generation and design guideline prompts.
	*
	* These are the canonical prompt fragments that instruct an LLM how to call
	* the render_a2ui tool, how to bind data, and how to style surfaces.
	*/
	/**
	* Generation guidelines — protocol rules, tool arguments, path rules,
	* data model format, and form/two-way-binding instructions.
	*/
	const A2UI_DEFAULT_GENERATION_GUIDELINES = `\
Generate A2UI v0.9 JSON.

## A2UI Protocol Instructions

A2UI (Agent to UI) is a protocol for rendering rich UI surfaces from agent responses.

CRITICAL: You MUST call the render_a2ui tool with ALL of these arguments:
- surfaceId: A unique ID for the surface (e.g. "product-comparison")
- components: REQUIRED — the A2UI component array. NEVER omit this. Only use
  components listed in the Available Components schema provided as context.
- data: OPTIONAL — a JSON object written to the root of the surface data model.
  Use for pre-filling form values or providing data for path-bound components.
- every component must have the "component" field specifying the component type.
  ONLY use component names from the Available Components schema — do NOT invent
  component names or use names not in the schema.

COMPONENT ID RULES:
- Exactly one component MUST have id="root". This is the surface's entry
  point — the renderer begins at "root" and walks the child/children tree
  from there. Every other component must be reachable from "root". If no
  component has id="root", the surface renders an empty loading placeholder
  and none of your components will be shown.
- Every component ID must be unique within the surface.
- A component MUST NOT reference itself as child/children. This causes a
  circular dependency error. For example, if a component has id="avatar",
  its child must be a DIFFERENT id (e.g. "avatar-img"), never "avatar".
- The child/children tree must be a DAG — no cycles allowed.

REPEATING CONTENT (TEMPLATES):
To repeat a component for each item in an array, use the structural children format:
  children: { componentId: "card-id", path: "/items" }
This tells the renderer to create one instance of "card-id" per item in the "/items" array.

PATH RULES FOR TEMPLATES:
Components inside a repeating template use RELATIVE paths (no leading slash).
The path is resolved relative to each array item automatically.
If a container has children: { componentId: "card", path: "/items" } and each item
has a key "name", use { "path": "name" } (NO leading slash — relative to item).
CRITICAL: Do NOT use "/name" (absolute) inside templates — use "name" (relative).
The container's path ("/items") uses a leading slash (absolute), but all
components INSIDE the template use paths WITHOUT leading slash.

COMPONENT VALUES — DEFAULT RULE:
Use inline literal values for ALL component properties. Pass strings, numbers,
arrays, and objects directly on the component. Do NOT use { "path": "..." }
objects unless the property's schema explicitly allows it (see exception below).
CRITICAL: USING { "path": "..." } ON A PROPERTY THAT DOES NOT DECLARE PATH
SUPPORT IN ITS SCHEMA WILL CAUSE A RUNTIME CRASH AND BREAK THE ENTIRE UI.
ALWAYS CHECK THE COMPONENT SCHEMA FIRST — IF THE PROPERTY ONLY ACCEPTS A
PLAIN TYPE, YOU MUST USE A LITERAL VALUE.
VERY IMPORTANT: THE APPLICATION WILL BREAK IF YOU DO NOT FOLLOW THIS RULE!

For example, a chart's "data" must always be an inline array:
  "data": [{"label": "Jan", "value": 100}, {"label": "Feb", "value": 200}]
A metric's "value" must always be an inline string:
  "value": "$1,200"

PATH BINDING EXCEPTION — SCHEMA-DRIVEN:
A few properties accept { "path": "/some/path" } as an alternative to a literal
value. You can identify these in the Available Components schema: the property
will list BOTH a literal type AND an object-with-path option. If a property only
shows a single type (string, number, array, etc.), it does NOT support path
binding — use a literal value only.

Path binding is typically used for editable form inputs so the client can write
user input back to the data model. When building forms:
- Bind input "value" to a data model path: "value": { "path": "/form/name" }
- Pre-fill via the "data" tool argument: "data": { "form": { "name": "Alice" } }
- Capture values on submit via button action context:
    "action": { "event": { "name": "submit", "context": { "name": { "path": "/form/name" } } } }

REPEATING CONTENT uses a structural children format (not the same as value binding):
  children: { componentId: "card-id", path: "/items" }
Components inside templates use RELATIVE paths (no leading slash): { "path": "name" }.`;
	/**
	* Design guidelines — visual design rules, component hierarchy tips,
	* and action handler patterns.
	*/
	const A2UI_DEFAULT_DESIGN_GUIDELINES = `\
Create polished, visually appealing interfaces. ONLY use components listed in the
Available Components schema — do NOT use component names that are not in the schema.

Design principles:
- Create clear visual hierarchy within cards and layouts.
- Keep cards clean — avoid clutter. Whitespace is good.
- Use consistent surfaceIds (lowercase, hyphenated).
- NEVER use the same ID for a component and its child — this creates a
  circular dependency. E.g. if id="avatar", child must NOT be "avatar".
- For side-by-side comparisons, use a container with structural children
  (children: { componentId, path }) to repeat a card template per data item.
- Include images when relevant (logos, icons, product photos):
  - Prefer company logos via Google favicons: https://www.google.com/s2/favicons?domain=example.com&sz=128
  - Do NOT invent Unsplash photo-IDs — they will 404. Only use real, known URLs.
- For buttons: action MUST use this exact nested format:
    "action": { "event": { "name": "myAction", "context": { "key": "value" } } }
  The "event" key holds an OBJECT with "name" (required) and "context" (optional).
  Do NOT use a flat format like {"event": "name"} — "event" must be an object.
- For forms: check the component schema — if an input's "value" property
  supports path binding, use it for editable fields. The submit button's
  action context should reference the same paths to capture user input.

Use the SAME surfaceId as the main surface. Match action names to button action event names.`;

//#endregion
//#region src/index.ts
	const COPILOTKIT_VERSION = version;
	/**
	* Client-safe license context factory.
	*
	* When status is null (no token provided), all features return true
	* (unlicensed = unrestricted, with branding). This is inlined here to
	* avoid importing the full license-verifier bundle (which depends on
	* Node's `crypto`) into browser bundles.
	*/
	function createLicenseContextValue(status) {
		return {
			status: null,
			license: null,
			checkFeature: () => true,
			getLimit: () => null
		};
	}

//#endregion
exports.A2UI_DEFAULT_DESIGN_GUIDELINES = A2UI_DEFAULT_DESIGN_GUIDELINES;
exports.A2UI_DEFAULT_GENERATION_GUIDELINES = A2UI_DEFAULT_GENERATION_GUIDELINES;
exports.AG_UI_CHANNEL_EVENT = AG_UI_CHANNEL_EVENT;
exports.BANNER_ERROR_NAMES = BANNER_ERROR_NAMES;
exports.COPILOTKIT_VERSION = COPILOTKIT_VERSION;
exports.COPILOT_CLOUD_API_URL = COPILOT_CLOUD_API_URL;
exports.COPILOT_CLOUD_CHAT_URL = COPILOT_CLOUD_CHAT_URL;
exports.COPILOT_CLOUD_ERROR_NAMES = COPILOT_CLOUD_ERROR_NAMES;
exports.COPILOT_CLOUD_PUBLIC_API_KEY_HEADER = COPILOT_CLOUD_PUBLIC_API_KEY_HEADER;
exports.COPILOT_CLOUD_VERSION = COPILOT_CLOUD_VERSION;
exports.ConfigurationError = ConfigurationError;
exports.ConsoleColors = ConsoleColors;
exports.ConsoleStyles = ConsoleStyles;
exports.CopilotKitAgentDiscoveryError = CopilotKitAgentDiscoveryError;
exports.CopilotKitApiDiscoveryError = CopilotKitApiDiscoveryError;
exports.CopilotKitError = CopilotKitError;
exports.CopilotKitErrorCode = CopilotKitErrorCode;
exports.CopilotKitLowLevelError = CopilotKitLowLevelError;
exports.CopilotKitMisuseError = CopilotKitMisuseError;
exports.CopilotKitRemoteEndpointDiscoveryError = CopilotKitRemoteEndpointDiscoveryError;
exports.CopilotKitVersionMismatchError = CopilotKitVersionMismatchError;
exports.DEFAULT_AGENT_ID = DEFAULT_AGENT_ID;
exports.ERROR_CONFIG = ERROR_CONFIG;
exports.ERROR_NAMES = ERROR_NAMES;
exports.ErrorVisibility = ErrorVisibility;
exports.MissingPublicApiKeyError = MissingPublicApiKeyError;
exports.RUNTIME_MODE_INTELLIGENCE = RUNTIME_MODE_INTELLIGENCE;
exports.RUNTIME_MODE_SSE = RUNTIME_MODE_SSE;
exports.ResolvedCopilotKitError = ResolvedCopilotKitError;
exports.Severity = Severity;
exports.TelemetryClient = TelemetryClient;
exports.TranscriptionErrorCode = TranscriptionErrorCode;
exports.TranscriptionErrors = TranscriptionErrors;
exports.UpgradeRequiredError = UpgradeRequiredError;
exports.actionParametersToJsonSchema = actionParametersToJsonSchema;
exports.convertJsonSchemaToZodSchema = convertJsonSchemaToZodSchema;
exports.copyToClipboard = copyToClipboard;
exports.createLicenseContextValue = createLicenseContextValue;
exports.dataToUUID = dataToUUID;
exports.ensureStructuredError = ensureStructuredError;
exports.exceedsMaxSize = exceedsMaxSize;
exports.executeConditions = executeConditions;
exports.finalizeRunEvents = finalizeRunEvents;
exports.formatFileSize = formatFileSize;
exports.generateVideoThumbnail = generateVideoThumbnail;
exports.getDocumentIcon = getDocumentIcon;
exports.getModalityFromMimeType = getModalityFromMimeType;
exports.getPossibleVersionMismatch = getPossibleVersionMismatch;
exports.getSourceUrl = getSourceUrl;
exports.getZodParameters = getZodParameters;
exports.isMacOS = isMacOS;
exports.isStructuredCopilotKitError = isStructuredCopilotKitError;
exports.isTelemetryDisabled = isTelemetryDisabled;
exports.isValidUUID = isValidUUID;
exports.jsonSchemaToActionParameters = jsonSchemaToActionParameters;
exports.logCopilotKitPlatformMessage = logCopilotKitPlatformMessage;
exports.logStyled = logStyled;
exports.logger = logger;
exports.matchesAcceptFilter = matchesAcceptFilter;
exports.parseJson = parseJson;
exports.partialJSONParse = partialJSONParse;
exports.phoenixExponentialBackoff = phoenixExponentialBackoff;
exports.publicApiKeyRequired = publicApiKeyRequired;
exports.randomId = randomId;
exports.randomUUID = randomUUID;
exports.readBody = readBody;
exports.readFileAsBase64 = readFileAsBase64;
exports.resolveDebugConfig = resolveDebugConfig;
exports.safeParseToolArgs = safeParseToolArgs;
exports.schemaToJsonSchema = schemaToJsonSchema;
exports.styledConsole = styledConsole;
exports.tryMap = tryMap;
});
