// @copilotkit/react-core/dist/v2/index.cjs
"use client";

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_copilotkit = require('../copilotkit-C3k13WZn.cjs');
require("./index.css");
let _copilotkit_a2ui_renderer = require("@copilotkit/a2ui-renderer");

exports.AudioRecorderError = require_copilotkit.AudioRecorderError;
Object.defineProperty(exports, 'CopilotChat', {
  enumerable: true,
  get: function () {
    return require_copilotkit.CopilotChat;
  }
});
exports.CopilotChatAssistantMessage = require_copilotkit.CopilotChatAssistantMessage_default;
exports.CopilotChatAttachmentQueue = require_copilotkit.CopilotChatAttachmentQueue;
exports.CopilotChatAttachmentRenderer = require_copilotkit.CopilotChatAttachmentRenderer;
exports.CopilotChatAudioRecorder = require_copilotkit.CopilotChatAudioRecorder;
exports.CopilotChatConfigurationProvider = require_copilotkit.CopilotChatConfigurationProvider;
exports.CopilotChatInput = require_copilotkit.CopilotChatInput_default;
exports.CopilotChatMessageView = require_copilotkit.CopilotChatMessageView;
exports.CopilotChatReasoningMessage = require_copilotkit.CopilotChatReasoningMessage_default;
exports.CopilotChatSuggestionPill = require_copilotkit.CopilotChatSuggestionPill;
exports.CopilotChatSuggestionView = require_copilotkit.CopilotChatSuggestionView;
exports.CopilotChatToggleButton = require_copilotkit.CopilotChatToggleButton;
exports.CopilotChatToggleButtonCloseIcon = require_copilotkit.DefaultCloseIcon;
exports.CopilotChatToggleButtonOpenIcon = require_copilotkit.DefaultOpenIcon;
exports.CopilotChatToolCallsView = require_copilotkit.CopilotChatToolCallsView;
exports.CopilotChatUserMessage = require_copilotkit.CopilotChatUserMessage_default;
exports.CopilotChatView = require_copilotkit.CopilotChatView_default;
exports.CopilotKit = require_copilotkit.CopilotKit;
exports.CopilotKitCoreReact = require_copilotkit.CopilotKitCoreReact;
exports.CopilotKitInspector = require_copilotkit.CopilotKitInspector;
exports.CopilotKitProvider = require_copilotkit.CopilotKitProvider;
Object.defineProperty(exports, 'CopilotModalHeader', {
  enumerable: true,
  get: function () {
    return require_copilotkit.CopilotModalHeader;
  }
});
exports.CopilotPopup = require_copilotkit.CopilotPopup;
Object.defineProperty(exports, 'CopilotPopupView', {
  enumerable: true,
  get: function () {
    return require_copilotkit.CopilotPopupView;
  }
});
exports.CopilotSidebar = require_copilotkit.CopilotSidebar;
Object.defineProperty(exports, 'CopilotSidebarView', {
  enumerable: true,
  get: function () {
    return require_copilotkit.CopilotSidebarView;
  }
});
exports.IntelligenceIndicator = require_copilotkit.IntelligenceIndicator;
exports.MCPAppsActivityContentSchema = require_copilotkit.MCPAppsActivityContentSchema;
exports.MCPAppsActivityRenderer = require_copilotkit.MCPAppsActivityRenderer;
exports.MCPAppsActivityType = require_copilotkit.MCPAppsActivityType;
exports.SandboxFunctionsContext = require_copilotkit.SandboxFunctionsContext;
exports.UseAgentUpdate = require_copilotkit.UseAgentUpdate;
exports.WildcardToolCallRender = require_copilotkit.WildcardToolCallRender;
Object.defineProperty(exports, 'a2uiDefaultTheme', {
  enumerable: true,
  get: function () {
    return _copilotkit_a2ui_renderer.defaultTheme;
  }
});
exports.createA2UIMessageRenderer = require_copilotkit.createA2UIMessageRenderer;
exports.defineToolCallRenderer = require_copilotkit.defineToolCallRenderer;
exports.useAgent = require_copilotkit.useAgent;
exports.useAgentContext = require_copilotkit.useAgentContext;
exports.useAttachments = require_copilotkit.useAttachments;
exports.useCapabilities = require_copilotkit.useCapabilities;
exports.useComponent = require_copilotkit.useComponent;
exports.useConfigureSuggestions = require_copilotkit.useConfigureSuggestions;
exports.useCopilotChatConfiguration = require_copilotkit.useCopilotChatConfiguration;
exports.useCopilotKit = require_copilotkit.useCopilotKit;
exports.useDefaultRenderTool = require_copilotkit.useDefaultRenderTool;
exports.useFrontendTool = require_copilotkit.useFrontendTool;
exports.useHumanInTheLoop = require_copilotkit.useHumanInTheLoop;
exports.useInterrupt = require_copilotkit.useInterrupt;
exports.useRenderActivityMessage = require_copilotkit.useRenderActivityMessage;
exports.useRenderCustomMessages = require_copilotkit.useRenderCustomMessages;
exports.useRenderTool = require_copilotkit.useRenderTool;
exports.useRenderToolCall = require_copilotkit.useRenderToolCall;
exports.useSandboxFunctions = require_copilotkit.useSandboxFunctions;
exports.useSuggestions = require_copilotkit.useSuggestions;
exports.useThreads = require_copilotkit.useThreads$1;
var _copilotkit_core = require("@copilotkit/core");
Object.keys(_copilotkit_core).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return _copilotkit_core[k]; }
  });
});

var _ag_ui_client = require("@ag-ui/client");
Object.keys(_ag_ui_client).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return _ag_ui_client[k]; }
  });
});
