// @tanstack/pacer/dist/async-queuer.js
import { parseFunctionOrValue } from "./utils.js";
import { emitChange, pacerEventClient } from "./event-client.js";
import { AsyncRetryer } from "./async-retryer.js";
import { Store } from "@tanstack/store";

//#region src/async-queuer.ts
function getDefaultAsyncQueuerState() {
	return {
		activeItems: [],
		addItemCount: 0,
		errorCount: 0,
		executeCount: 0,
		expirationCount: 0,
		isEmpty: true,
		isExecuting: false,
		isFull: false,
		isIdle: true,
		isRunning: true,
		itemTimestamps: [],
		items: [],
		lastResult: null,
		pendingTick: false,
		rejectionCount: 0,
		settledCount: 0,
		size: 0,
		status: "idle",
		successCount: 0
	};
}
/**
* Utility function for sharing common `AsyncQueuerOptions` options between different `AsyncQueuer` instances.
*/
function asyncQueuerOptions(options) {
	return options;
}
const defaultOptions = {
	addItemsTo: "back",
	asyncRetryerOptions: { maxAttempts: 1 },
	concurrency: 1,
	expirationDuration: Infinity,
	getIsExpired: () => false,
	getItemsFrom: "front",
	getPriority: (item) => item?.priority ?? 0,
	initialItems: [],
	maxSize: Infinity,
	started: true,
	wait: 0
};
/**
* A flexible asynchronous queue for processing tasks with configurable concurrency, priority, and expiration.
*
* Async vs Sync Versions:
* The async version provides advanced features over the sync Queuer:
* - Returns promises that can be awaited for task results
* - Built-in retry support via AsyncRetryer integration for each queued task
* - Abort support to cancel in-flight task executions
* - Comprehensive error handling with onError callbacks and throwOnError control
* - Detailed execution tracking (success/error/settle counts)
* - Concurrent execution support (process multiple items simultaneously)
*
* The sync Queuer is lighter weight and simpler when you don't need async features,
* return values, or execution control.
*
* What is Queuing?
* Queuing is a technique for managing and processing items sequentially or with controlled concurrency.
* Tasks are processed up to the configured concurrency limit. When a task completes,
* the next pending task is processed if the concurrency limit allows.
*
* Key Features:
* - Priority queue support via the getPriority option
* - Configurable concurrency limit
* - Callbacks for task success, error, completion, and queue state changes
* - FIFO (First In First Out) or LIFO (Last In First Out) queue behavior
* - Pause and resume processing
* - Item expiration to remove stale items from the queue
*
* Error Handling:
* - If an `onError` handler is provided, it will be called with the error and queuer instance
* - If `throwOnError` is true (default when no onError handler is provided), the error will be thrown
* - If `throwOnError` is false (default when onError handler is provided), the error will be swallowed
* - Both onError and throwOnError can be used together; the handler will be called before any error is thrown
* - The error state can be checked using the AsyncQueuer instance
*
* State Management:
* - Uses TanStack Store for reactive state management
* - Use `initialState` to provide initial state values when creating the async queuer
* - Use `onSuccess` callback to react to successful task execution and implement custom logic
* - Use `onError` callback to react to task execution errors and implement custom error handling
* - Use `onSettled` callback to react to task execution completion (success or error) and implement custom logic
* - Use `onItemsChange` callback to react to items being added or removed from the queue
* - Use `onExpire` callback to react to items expiring and implement custom logic
* - Use `onReject` callback to react to items being rejected when the queue is full
* - The state includes error count, expiration count, rejection count, running status, and success/settle counts
* - State can be accessed via `asyncQueuer.store.state` when using the class directly
* - When using framework adapters (React/Solid), state is accessed from `asyncQueuer.state`
*
* Example usage:
* ```ts
* const asyncQueuer = new AsyncQueuer<string>(async (item) => {
*   // process item
*   return item.toUpperCase();
* }, {
*   concurrency: 2,
*   onSuccess: (result) => {
*     console.log(result);
*   }
* });
*
* asyncQueuer.addItem('hello');
* asyncQueuer.start();
* ```
*/
var AsyncQueuer = class {
	#timeoutIds = /* @__PURE__ */ new Set();
	constructor(fn, initialOptions = {}) {
		this.fn = fn;
		this.store = new Store(getDefaultAsyncQueuerState());
		this.asyncRetryers = /* @__PURE__ */ new Map();
		this.setOptions = (newOptions) => {
			this.options = {
				...this.options,
				...newOptions
			};
		};
		this.addItem = (item, position = this.options.addItemsTo ?? "back", runOnItemsChange = true) => {
			this.#setState({ addItemCount: this.store.state.addItemCount + 1 });
			if (this.store.state.items.length >= (this.options.maxSize ?? Infinity)) {
				this.#setState({ rejectionCount: this.store.state.rejectionCount + 1 });
				this.options.onReject?.(item, this);
				return false;
			}
			const priority = this.options.getPriority !== defaultOptions.getPriority ? this.options.getPriority(item) : item.priority;
			const items = this.store.state.items;
			const itemTimestamps = this.store.state.itemTimestamps;
			if (priority !== void 0) {
				const insertIndex = items.findIndex((existing) => {
					return (this.options.getPriority !== defaultOptions.getPriority ? this.options.getPriority(existing) : existing.priority) < priority;
				});
				if (insertIndex === -1) {
					items.push(item);
					itemTimestamps.push(Date.now());
				} else {
					items.splice(insertIndex, 0, item);
					itemTimestamps.splice(insertIndex, 0, Date.now());
				}
			} else if (position === "front") {
				items.unshift(item);
				itemTimestamps.unshift(Date.now());
			} else {
				items.push(item);
				itemTimestamps.push(Date.now());
			}
			this.#setState({
				items,
				itemTimestamps
			});
			if (runOnItemsChange) this.options.onItemsChange?.(this);
			if (this.store.state.isRunning && !this.store.state.pendingTick) this.#tick();
			return true;
		};
		this.getNextItem = (position = this.options.getItemsFrom ?? "front") => {
			const { items, itemTimestamps } = this.store.state;
			let item;
			if (this.options.getPriority !== defaultOptions.getPriority || position === "front") {
				item = items[0];
				if (item !== void 0) this.#setState({
					items: items.slice(1),
					itemTimestamps: itemTimestamps.slice(1)
				});
			} else {
				item = items[items.length - 1];
				if (item !== void 0) this.#setState({
					items: items.slice(0, -1),
					itemTimestamps: itemTimestamps.slice(0, -1)
				});
			}
			if (item !== void 0) this.options.onItemsChange?.(this);
			return item;
		};
		this.execute = async (position) => {
			const item = this.getNextItem(position);
			if (item !== void 0) {
				const currentExecuteCount = this.store.state.executeCount + 1;
				this.#setState({
					executeCount: currentExecuteCount,
					isExecuting: true
				});
				try {
					const currentAsyncRetryer = new AsyncRetryer(this.fn, {
						...this.options.asyncRetryerOptions,
						key: `${this.key}-retryer-${currentExecuteCount}`
					});
					this.asyncRetryers.set(currentExecuteCount, currentAsyncRetryer);
					const lastResult = await currentAsyncRetryer.execute(item);
					this.#setState({
						successCount: this.store.state.successCount + 1,
						lastResult
					});
					this.options.onSuccess?.(lastResult, item, this);
				} catch (error) {
					this.#setState({ errorCount: this.store.state.errorCount + 1 });
					this.options.onError?.(error, item, this);
					if (this.options.throwOnError) throw error;
				} finally {
					this.asyncRetryers.delete(currentExecuteCount);
					this.#setState({
						activeItems: this.store.state.activeItems.filter((activeItem) => activeItem !== item),
						isExecuting: false,
						settledCount: this.store.state.settledCount + 1
					});
					this.options.onSettled?.(item, this);
				}
			}
			return item;
		};
		this.flush = async (numberOfItems = this.store.state.items.length, position) => {
			this.#clearTimeouts();
			await Promise.all(Array.from({ length: numberOfItems }, () => this.execute(position)));
		};
		this.flushAsBatch = async (batchFunction) => {
			this.#clearTimeouts();
			await batchFunction(this.#getAllItems());
		};
		this.peekNextItem = (position = "front") => {
			if (position === "front") return this.store.state.items[0];
			return this.store.state.items[this.store.state.items.length - 1];
		};
		this.peekAllItems = () => {
			return [...this.peekActiveItems(), ...this.peekPendingItems()];
		};
		this.peekActiveItems = () => {
			return [...this.store.state.activeItems];
		};
		this.peekPendingItems = () => {
			return [...this.store.state.items];
		};
		this.start = () => {
			this.#setState({ isRunning: true });
			if (!this.store.state.pendingTick && this.store.state.items.length > 0) this.#tick();
		};
		this.stop = () => {
			this.#clearTimeouts();
			this.#setState({
				isRunning: false,
				pendingTick: false
			});
		};
		this.clear = () => {
			this.#setState({
				items: [],
				itemTimestamps: []
			});
			this.options.onItemsChange?.(this);
		};
		this.getAbortSignal = (executeCount) => {
			const count = executeCount ?? this.store.state.executeCount;
			return this.asyncRetryers.get(count)?.getAbortSignal() ?? null;
		};
		this.abort = () => {
			this.asyncRetryers.forEach((retryer) => retryer.abort());
			this.asyncRetryers.clear();
			this.#setState({ isExecuting: false });
		};
		this.reset = () => {
			this.#setState(getDefaultAsyncQueuerState());
			this.options.onItemsChange?.(this);
			this.asyncRetryers.forEach((retryer) => retryer.reset());
		};
		this.key = initialOptions.key;
		this.options = {
			...defaultOptions,
			...initialOptions,
			throwOnError: initialOptions.throwOnError ?? !initialOptions.onError
		};
		const isInitiallyRunning = this.options.initialState?.isRunning ?? this.options.started ?? true;
		this.#setState({
			...this.options.initialState,
			isRunning: isInitiallyRunning
		});
		if (this.options.initialState?.items) {
			if (this.store.state.isRunning) this.#tick();
		} else for (let i = 0; i < (this.options.initialItems?.length ?? 0); i++) {
			const item = this.options.initialItems[i];
			const isLast = i === (this.options.initialItems?.length ?? 0) - 1;
			this.addItem(item, this.options.addItemsTo ?? "back", isLast);
		}
		if (this.key) pacerEventClient.on("d-AsyncQueuer", (e) => {
			if (e.payload.key !== this.key) return;
			this.#setState(e.payload.store.state);
			this.setOptions(e.payload.options);
		});
	}
	#setState = (newState) => {
		this.store.setState((state) => {
			const combinedState = {
				...state,
				...newState
			};
			const { activeItems, items, isRunning } = combinedState;
			const size = items.length;
			const isFull = size >= (this.options.maxSize ?? Infinity);
			const isEmpty = size === 0;
			const isIdle = isRunning && isEmpty && activeItems.length === 0;
			const status = isIdle ? "idle" : isRunning ? "running" : "stopped";
			return {
				...combinedState,
				isEmpty,
				isFull,
				isIdle,
				size,
				status
			};
		});
		emitChange("AsyncQueuer", this);
	};
	/**
	* Returns the current wait time (in milliseconds) between processing items.
	* If a function is provided, it is called with the queuer instance.
	*/
	#getWait = () => {
		return parseFunctionOrValue(this.options.wait ?? 0, this);
	};
	/**
	* Returns the current concurrency limit for processing items.
	* If a function is provided, it is called with the queuer instance.
	*/
	#getConcurrency = () => {
		return parseFunctionOrValue(this.options.concurrency ?? 1, this);
	};
	/**
	* Processes items in the queue up to the concurrency limit. Internal use only.
	*/
	#tick = () => {
		if (!this.store.state.isRunning) {
			this.#setState({ pendingTick: false });
			return;
		}
		this.#setState({ pendingTick: true });
		this.#checkExpiredItems();
		const activeItems = this.store.state.activeItems;
		while (activeItems.length < this.#getConcurrency() && this.store.state.items.length > 0) {
			const nextItem = this.peekNextItem();
			if (!nextItem) break;
			activeItems.push(nextItem);
			this.#setState({ activeItems });
			(async () => {
				await this.execute();
				const wait = this.#getWait();
				if (wait > 0) {
					const timeoutId = setTimeout(() => this.#tick(), wait);
					this.#timeoutIds.add(timeoutId);
					return;
				}
				this.#tick();
			})();
		}
		this.#setState({ pendingTick: false });
	};
	#getAllItems = () => {
		const items = this.peekAllItems();
		this.clear();
		return items;
	};
	/**
	* Checks for expired items in the queue and removes them. Calls onExpire for each expired item.
	* Internal use only.
	*/
	#checkExpiredItems = () => {
		if ((this.options.expirationDuration ?? Infinity) === Infinity && this.options.getIsExpired === defaultOptions.getIsExpired) return;
		const now = Date.now();
		const expiredIndices = [];
		for (let i = 0; i < this.store.state.items.length; i++) {
			const timestamp = this.store.state.itemTimestamps[i];
			if (timestamp === void 0) continue;
			const item = this.store.state.items[i];
			if (item === void 0) continue;
			if (this.options.getIsExpired !== defaultOptions.getIsExpired ? this.options.getIsExpired(item, timestamp) : now - timestamp > (this.options.expirationDuration ?? Infinity)) expiredIndices.push(i);
		}
		for (let i = expiredIndices.length - 1; i >= 0; i--) {
			const index = expiredIndices[i];
			if (index === void 0) continue;
			const expiredItem = this.store.state.items[index];
			if (expiredItem === void 0) continue;
			const newItems = [...this.store.state.items];
			const newTimestamps = [...this.store.state.itemTimestamps];
			newItems.splice(index, 1);
			newTimestamps.splice(index, 1);
			this.#setState({
				items: newItems,
				itemTimestamps: newTimestamps,
				expirationCount: this.store.state.expirationCount + 1
			});
			this.options.onExpire?.(expiredItem, this);
		}
		if (expiredIndices.length > 0) this.options.onItemsChange?.(this);
	};
	#clearTimeouts = () => {
		this.#timeoutIds.forEach((timeoutId) => clearTimeout(timeoutId));
		this.#timeoutIds.clear();
	};
};
/**
* Creates a new AsyncQueuer instance and returns a bound addItem function for adding tasks.
* The queuer is started automatically and ready to process items.
*
* Async vs Sync Versions:
* The async version provides advanced features over the sync queue function:
* - Returns promises that can be awaited for task results
* - Built-in retry support via AsyncRetryer integration for each queued task
* - Abort support to cancel in-flight task executions
* - Comprehensive error handling with onError callbacks and throwOnError control
* - Detailed execution tracking (success/error/settle counts)
* - Concurrent execution support (process multiple items simultaneously)
*
* The sync queue function is lighter weight and simpler when you don't need async features,
* return values, or execution control.
*
* What is Queuing?
* Queuing is a technique for managing and processing items sequentially or with controlled concurrency.
* Tasks are processed up to the configured concurrency limit. When a task completes,
* the next pending task is processed if the concurrency limit allows.
*
* Configuration Options:
* - `concurrency`: Maximum number of concurrent tasks (default: 1)
* - `wait`: Time to wait between processing items (default: 0)
* - `maxSize`: Maximum number of items allowed in the queue (default: Infinity)
* - `getPriority`: Function to determine item priority
* - `addItemsTo`: Default position to add items ('back' or 'front', default: 'back')
* - `getItemsFrom`: Default position to get items ('front' or 'back', default: 'front')
* - `expirationDuration`: Maximum time items can stay in queue
* - `started`: Whether to start processing immediately (default: true)
* - `asyncRetryerOptions`: Configure retry behavior for task executions
*
* Error Handling:
* - If an `onError` handler is provided, it will be called with the error and queuer instance
* - If `throwOnError` is true (default when no onError handler is provided), the error will be thrown
* - If `throwOnError` is false (default when onError handler is provided), the error will be swallowed
* - Both onError and throwOnError can be used together; the handler will be called before any error is thrown
* - The error state can be checked using the underlying AsyncQueuer instance
*
* State Management:
* - Uses TanStack Store for reactive state management
* - Use `initialState` to provide initial state values when creating the async queuer
* - Use `onSuccess` callback to react to successful task execution and implement custom logic
* - Use `onError` callback to react to task execution errors and implement custom error handling
* - Use `onSettled` callback to react to task execution completion (success or error) and implement custom logic
* - Use `onItemsChange` callback to react to items being added or removed from the queue
* - Use `onExpire` callback to react to items expiring and implement custom logic
* - Use `onReject` callback to react to items being rejected when the queue is full
* - The state includes error count, expiration count, rejection count, running status, and success/settle counts
* - State can be accessed via the underlying AsyncQueuer instance's `store.state` property
* - When using framework adapters (React/Solid), state is accessed from the hook's state property
*
* @example
* ```ts
* const enqueue = asyncQueue<string>(async (item) => {
*   return item.toUpperCase();
* }, {
*   concurrency: 2,
*   wait: 100,
*   onSuccess: (result) => console.log('Processed:', result)
* });
*
* enqueue('hello');
* ```
*/
function asyncQueue(fn, initialOptions) {
	return new AsyncQueuer(fn, initialOptions).addItem;
}

//#endregion
export { AsyncQueuer, asyncQueue, asyncQueuerOptions };
