// @tanstack/pacer/dist/async-rate-limiter.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_utils = require('./utils.cjs');
const require_event_client = require('./event-client.cjs');
const require_async_retryer = require('./async-retryer.cjs');
let _tanstack_store = require("@tanstack/store");

//#region src/async-rate-limiter.ts
function getDefaultAsyncRateLimiterState() {
	return {
		errorCount: 0,
		executionTimes: [],
		isExceeded: false,
		isExecuting: false,
		lastResult: void 0,
		maybeExecuteCount: 0,
		rejectionCount: 0,
		settleCount: 0,
		status: "idle",
		successCount: 0
	};
}
/**
* Utility function for sharing common `AsyncRateLimiterOptions` options between different `AsyncRateLimiter` instances.
*/
function asyncRateLimiterOptions(options) {
	return options;
}
const defaultOptions = {
	asyncRetryerOptions: { maxAttempts: 1 },
	enabled: true,
	limit: 1,
	window: 0,
	windowType: "fixed",
	throwOnError: true
};
/**
* A class that creates an async rate-limited function.
*
* Async vs Sync Versions:
* The async version provides advanced features over the sync RateLimiter:
* - Returns promises that can be awaited for rate-limited function results
* - Built-in retry support via AsyncRetryer integration
* - Abort support to cancel in-flight executions
* - Comprehensive error handling with onError callbacks and throwOnError control
* - Detailed execution tracking (success/error/settle counts, rejection counts)
* - More sophisticated window management with automatic cleanup
*
* The sync RateLimiter is lighter weight and simpler when you don't need async features,
* return values, or execution control.
*
* What is Rate Limiting?
* Rate limiting allows a function to execute up to a limit within a time window,
* then blocks all subsequent calls until the window passes. This can lead to "bursty" behavior where
* all executions happen immediately, followed by a complete block.
*
* Window Types:
* - 'fixed': A strict window that resets after the window period. All executions within the window count
*   towards the limit, and the window resets completely after the period.
* - 'sliding': A rolling window that allows executions as old ones expire. This provides a more
*   consistent rate of execution over time.
*
* When to Use Rate Limiting:
* Rate limiting is best used for hard API limits or resource constraints. For UI updates or
* smoothing out frequent events, throttling or debouncing usually provide better user experience.
* - Throttling: Ensures consistent spacing between executions (e.g. max once per 200ms)
* - Debouncing: Waits for a pause in calls before executing (e.g. after 500ms of no calls)
*
* State Management:
* - Uses TanStack Store for reactive state management
* - Use `initialState` to provide initial state values when creating the rate limiter
* - `initialState` can be a partial state object
* - Use `onSuccess` callback to react to successful function execution and implement custom logic
* - Use `onError` callback to react to function execution errors and implement custom error handling
* - Use `onSettled` callback to react to function execution completion (success or error) and implement custom logic
* - Use `onReject` callback to react to executions being rejected when rate limit is exceeded
* - The state includes execution times, success/error counts, and current execution status
* - State can be accessed via `asyncRateLimiter.store.state` when using the class directly
* - When using framework adapters (React/Solid), state is accessed from `asyncRateLimiter.state`
*
* Error Handling:
* - If an `onError` handler is provided, it will be called with the error and rate limiter instance
* - If `throwOnError` is true (default when no onError handler is provided), the error will be thrown
* - If `throwOnError` is false (default when onError handler is provided), the error will be swallowed
* - Both onError and throwOnError can be used together - the handler will be called before any error is thrown
* - The error state can be checked using the underlying AsyncRateLimiter instance
* - Rate limit rejections (when limit is exceeded) are handled separately from execution errors via the `onReject` handler
*
* @example
* ```ts
* const rateLimiter = new AsyncRateLimiter(
*   async (id: string) => await api.getData(id),
*   {
*     limit: 5,
*     window: 1000,
*     windowType: 'sliding',
*     onError: (error) => {
*       console.error('API call failed:', error);
*     },
*     onReject: (limiter) => {
*       console.log(`Rate limit exceeded. Try again in ${limiter.getMsUntilNextWindow()}ms`);
*     }
*   }
* );
*
* // Will execute immediately until limit reached, then block
* // Returns the API response directly
* const data = await rateLimiter.maybeExecute('123');
* ```
*/
var AsyncRateLimiter = class {
	#timeoutIds = /* @__PURE__ */ new Set();
	constructor(fn, initialOptions) {
		this.fn = fn;
		this.store = new _tanstack_store.Store(getDefaultAsyncRateLimiterState());
		this.asyncRetryers = /* @__PURE__ */ new Map();
		this.setOptions = (newOptions) => {
			this.options = {
				...this.options,
				...newOptions
			};
		};
		this.maybeExecute = async (...args) => {
			this.#setState({ maybeExecuteCount: this.store.state.maybeExecuteCount + 1 });
			this.#cleanupOldExecutions();
			if (this.#getExecutionTimesInWindow().length < this.#getLimit()) {
				await this.#execute(...args);
				return this.store.state.lastResult;
			}
			this.#setState({ rejectionCount: this.store.state.rejectionCount + 1 });
			this.options.onReject?.(args, this);
		};
		this.getRemainingInWindow = () => {
			const relevantExecutionTimes = this.#getExecutionTimesInWindow();
			return Math.max(0, this.#getLimit() - relevantExecutionTimes.length);
		};
		this.getMsUntilNextWindow = () => {
			if (this.getRemainingInWindow() > 0) return 0;
			return (this.store.state.executionTimes[0] ?? Infinity) + this.#getWindow() - Date.now();
		};
		this.getAbortSignal = (maybeExecuteCount) => {
			const count = maybeExecuteCount ?? this.store.state.maybeExecuteCount;
			return this.asyncRetryers.get(count)?.getAbortSignal() ?? null;
		};
		this.abort = () => {
			this.asyncRetryers.forEach((retryer) => retryer.abort());
			this.asyncRetryers.clear();
			this.#setState({ isExecuting: false });
		};
		this.reset = () => {
			this.#setState(getDefaultAsyncRateLimiterState());
			this.#clearTimeouts();
			this.asyncRetryers.forEach((retryer) => retryer.reset());
		};
		this.key = initialOptions.key;
		this.options = {
			...defaultOptions,
			...initialOptions,
			throwOnError: initialOptions.throwOnError ?? !initialOptions.onError
		};
		this.#setState(this.options.initialState ?? {});
		for (const executionTime of this.#getExecutionTimesInWindow()) this.#setCleanupTimeout(executionTime);
		if (this.key) require_event_client.pacerEventClient.on("d-AsyncRateLimiter", (event) => {
			if (event.payload.key !== this.key) return;
			this.#setState(event.payload.store.state);
			this.setOptions(event.payload.options);
		});
	}
	#setState = (newState) => {
		this.store.setState((state) => {
			const combinedState = {
				...state,
				...newState
			};
			const isExceeded = combinedState.executionTimes.length >= this.#getLimit();
			const status = !this.#getEnabled() ? "disabled" : combinedState.isExecuting ? "executing" : isExceeded ? "exceeded" : "idle";
			return {
				...combinedState,
				isExceeded,
				status
			};
		});
		require_event_client.emitChange("AsyncRateLimiter", this);
	};
	/**
	* Returns the current enabled state of the async rate limiter
	*/
	#getEnabled = () => {
		return !!require_utils.parseFunctionOrValue(this.options.enabled, this);
	};
	/**
	* Returns the current limit of executions allowed within the time window
	*/
	#getLimit = () => {
		return require_utils.parseFunctionOrValue(this.options.limit, this);
	};
	/**
	* Returns the current time window in milliseconds
	*/
	#getWindow = () => {
		return require_utils.parseFunctionOrValue(this.options.window, this);
	};
	#execute = async (...args) => {
		if (!this.#getEnabled()) return;
		const currentMaybeExecute = this.store.state.maybeExecuteCount;
		const now = Date.now();
		const executionTimes = [...this.store.state.executionTimes, now];
		this.#setState({
			isExecuting: true,
			executionTimes
		});
		try {
			const currentAsyncRetryer = new require_async_retryer.AsyncRetryer(this.fn, {
				...this.options.asyncRetryerOptions,
				key: `${this.key}-retryer-${currentMaybeExecute}`
			});
			this.asyncRetryers.set(currentMaybeExecute, currentAsyncRetryer);
			const result = await currentAsyncRetryer.execute(...args);
			this.#setCleanupTimeout(now);
			this.#setState({
				successCount: this.store.state.successCount + 1,
				lastResult: result
			});
			this.options.onSuccess?.(result, args, this);
		} catch (error) {
			this.#setState({ errorCount: this.store.state.errorCount + 1 });
			this.options.onError?.(error, args, this);
			if (this.options.throwOnError) throw error;
		} finally {
			this.asyncRetryers.delete(currentMaybeExecute);
			this.#setState({
				isExecuting: false,
				settleCount: this.store.state.settleCount + 1
			});
			this.options.onSettled?.(args, this);
		}
		return this.store.state.lastResult;
	};
	#getExecutionTimesInWindow = () => {
		if (this.options.windowType === "sliding") return this.store.state.executionTimes.filter((time) => time > Date.now() - this.#getWindow());
		else {
			if (this.store.state.executionTimes.length === 0) return [];
			const windowStart = Math.min(...this.store.state.executionTimes);
			const windowEnd = windowStart + this.#getWindow();
			if (Date.now() > windowEnd) return [];
			return this.store.state.executionTimes.filter((time) => time >= windowStart && time <= windowEnd);
		}
	};
	#setCleanupTimeout = (executionTime) => {
		if (this.options.windowType === "sliding" || this.#timeoutIds.size === 0) {
			const timeUntilExpiration = executionTime - Date.now() + this.#getWindow() + 1;
			const timeoutId = setTimeout(() => {
				this.#cleanupOldExecutions();
				this.#clearTimeout(timeoutId);
			}, timeUntilExpiration);
			this.#timeoutIds.add(timeoutId);
		}
	};
	#clearTimeout = (timeoutId) => {
		clearTimeout(timeoutId);
		this.#timeoutIds.delete(timeoutId);
	};
	#clearTimeouts = () => {
		this.#timeoutIds.forEach((timeoutId) => clearTimeout(timeoutId));
		this.#timeoutIds.clear();
	};
	#cleanupOldExecutions = () => {
		this.#setState({ executionTimes: this.#getExecutionTimesInWindow() });
	};
};
/**
* Creates an async rate-limited function that will execute the provided function up to a maximum number of times within a time window.
*
* Async vs Sync Versions:
* The async version provides advanced features over the sync rate limit function:
* - Returns promises that can be awaited for rate-limited function results
* - Built-in retry support via AsyncRetryer integration
* - Abort support to cancel in-flight executions
* - Comprehensive error handling with onError callbacks and throwOnError control
* - Detailed execution tracking (success/error/settle counts, rejection counts)
* - More sophisticated window management with automatic cleanup
*
* The sync rate limit function is lighter weight and simpler when you don't need async features,
* return values, or execution control.
*
* What is Rate Limiting?
* Rate limiting allows a function to execute up to a limit within a time window,
* then blocks all subsequent calls until the window passes. This can lead to "bursty" behavior where
* all executions happen immediately, followed by a complete block.
*
* Window Types:
* - 'fixed': A strict window that resets after the window period. All executions within the window count
*   towards the limit, and the window resets completely after the period.
* - 'sliding': A rolling window that allows executions as old ones expire. This provides a more
*   consistent rate of execution over time.
*
* Configuration Options:
* - `limit`: Maximum number of executions allowed within the window (required)
* - `window`: Time window in milliseconds (required)
* - `windowType`: 'fixed' or 'sliding' (default: 'fixed')
* - `enabled`: Whether the rate limiter is enabled (default: true)
* - `asyncRetryerOptions`: Configure retry behavior for executions
*
* When to Use Rate Limiting:
* Rate limiting is best used for hard API limits or resource constraints. For UI updates or
* smoothing out frequent events, throttling or debouncing usually provide better user experience.
* - A rate limiter will allow all executions until the limit is reached, then block all subsequent calls until the window resets
* - A throttler ensures even spacing between executions, which can be better for consistent performance
* - A debouncer collapses multiple calls into one, which is better for handling bursts of events
*
* Error Handling:
* - If an `onError` handler is provided, it will be called with the error and rate limiter instance
* - If `throwOnError` is true (default when no onError handler is provided), the error will be thrown
* - If `throwOnError` is false (default when onError handler is provided), the error will be swallowed
* - Both onError and throwOnError can be used together - the handler will be called before any error is thrown
* - The error state can be checked using the underlying AsyncRateLimiter instance
* - Rate limit rejections (when limit is exceeded) are handled separately from execution errors via the `onReject` handler
*
* State Management:
* - Uses TanStack Store for reactive state management
* - Use `initialState` to provide initial state values when creating the rate limiter
* - `initialState` can be a partial state object
* - Use `onSuccess` callback to react to successful function execution and implement custom logic
* - Use `onError` callback to react to function execution errors and implement custom error handling
* - Use `onSettled` callback to react to function execution completion (success or error) and implement custom logic
* - Use `onReject` callback to react to executions being rejected when rate limit is exceeded
* - The state includes execution times, success/error counts, and current execution status
* - State can be accessed via the underlying AsyncRateLimiter instance's `store.state` property
* - When using framework adapters (React/Solid), state is accessed from the hook's state property
*
* @example
* ```ts
* // Rate limit to 5 calls per minute with a sliding window
* const rateLimited = asyncRateLimit(makeApiCall, {
*   limit: 5,
*   window: 60000,
*   windowType: 'sliding',
*   onError: (error) => {
*     console.error('API call failed:', error);
*   },
*   onReject: (rateLimiter) => {
*     console.log(`Rate limit exceeded. Try again in ${rateLimiter.getMsUntilNextWindow()}ms`);
*   }
* });
*
* // First 5 calls will execute immediately
* // Additional calls will be rejected until the minute window resets
* // Returns the API response directly
* const result = await rateLimited();
*
* // For more even execution, consider using throttle instead:
* const throttled = throttle(makeApiCall, { wait: 12000 }); // One call every 12 seconds
* ```
*/
function asyncRateLimit(fn, initialOptions) {
	return new AsyncRateLimiter(fn, initialOptions).maybeExecute;
}

//#endregion
exports.AsyncRateLimiter = AsyncRateLimiter;
exports.asyncRateLimit = asyncRateLimit;
exports.asyncRateLimiterOptions = asyncRateLimiterOptions;
