// @tanstack/pacer/dist/async-retryer.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_utils = require('./utils.cjs');
const require_event_client = require('./event-client.cjs');
let _tanstack_store = require("@tanstack/store");

//#region src/async-retryer.ts
/**
* Creates the default initial state for an AsyncRetryer instance
* @returns The default state with all values reset to initial values
*/
function getDefaultAsyncRetryerState() {
	return {
		currentAttempt: 0,
		executionCount: 0,
		isExecuting: false,
		lastError: void 0,
		lastExecutionTime: 0,
		lastResult: void 0,
		status: "idle",
		totalExecutionTime: 0
	};
}
/**
* Utility function for sharing common `AsyncRetryerOptions` options between different `AsyncRetryer` instances.
*/
function asyncRetryerOptions(options) {
	return options;
}
const defaultOptions = {
	backoff: "exponential",
	baseWait: 1e3,
	maxWait: Infinity,
	enabled: true,
	jitter: 0,
	maxAttempts: 3,
	maxExecutionTime: Infinity,
	maxTotalExecutionTime: Infinity,
	throwOnError: "last"
};
/**
* Provides robust retry functionality for asynchronous functions, supporting configurable backoff strategies,
* attempt limits, timeout controls, and detailed state management. The AsyncRetryer class is designed to help you reliably
* execute async operations that may fail intermittently, such as network requests or database operations,
* by automatically retrying them according to your chosen policy.
*
* ## Retrying Concepts
*
* - **Retrying**: Automatically re-executes a failed async function up to a specified number of attempts.
*   Useful for handling transient errors (e.g., network flakiness, rate limits, temporary server issues).
* - **Backoff Strategies**: Controls the delay between retry attempts (default: `'exponential'`):
*   - `'exponential'`: Wait time doubles with each attempt (1s, 2s, 4s, ...) - **DEFAULT**
*   - `'linear'`: Wait time increases linearly (1s, 2s, 3s, ...)
*   - `'fixed'`: Waits a constant amount of time (`baseWait`) between each attempt
* - **Jitter**: Adds randomness to retry delays to prevent thundering herd problems (default: `0`).
*   Set to a value between 0-1 to apply that percentage of random variation to each delay.
* - **Max Wait**: Caps the maximum wait time between retries (default: `Infinity`).
*   Useful for preventing exponential backoff from growing too large (e.g., cap at 30s even if exponential would be 64s).
* - **Timeout Controls**: Set limits on execution time to prevent hanging operations:
*   - `maxExecutionTime`: Maximum time for a single function call (default: `Infinity`)
*   - `maxTotalExecutionTime`: Maximum time for the entire retry operation (default: `Infinity`)
* - **Abort & Cancellation**: Supports cancellation via an internal `AbortController`. Call `abort()` to stop retries.
*   Use `getAbortSignal()` to make your async function actually cancellable (e.g., with fetch requests).
*
* ## State Management
*
* Uses TanStack Store for fine-grained reactivity. State can be accessed via the `store.state` property.
*
* Available state properties:
* - `currentAttempt`: The current retry attempt number (0 when not executing)
* - `executionCount`: Total number of completed executions (successful or failed)
* - `isExecuting`: Whether the retryer is currently executing the function
* - `lastError`: The most recent error encountered during execution
* - `lastExecutionTime`: Timestamp of the last execution completion in milliseconds
* - `lastResult`: The result from the most recent successful execution
* - `status`: Current execution status ('disabled' | 'idle' | 'executing' | 'retrying')
* - `totalExecutionTime`: Total time spent executing (including retries) in milliseconds
*
* ## Error Handling
*
* The `throwOnError` option controls when errors are thrown (default: `'last'`):
* - `'last'`: Only throws the final error after all retries are exhausted - **DEFAULT**
* - `true`: Throws every error immediately (disables retrying)
* - `false`: Never throws errors, returns `undefined` instead
*
* Callbacks for lifecycle management:
* - `onAbort`: Called when execution is aborted (manually or due to timeouts)
* - `onError`: Called for every error (including during retries)
* - `onLastError`: Called only for the final error after all retries fail
* - `onRetry`: Called before each retry attempt
* - `onSettled`: Called after execution completes (success or failure) of each attempt
* - `onSuccess`: Called when execution succeeds
* - `onExecutionTimeout`: Called when a single execution attempt times out
* - `onTotalExecutionTimeout`: Called when the total execution time times out
*
* ## Usage
*
* - Use for async operations that may fail transiently and benefit from retrying.
* - Configure `maxAttempts`, `backoff`, `baseWait`, `maxWait`, and `jitter` to control retry behavior.
* - Set `maxExecutionTime` and `maxTotalExecutionTime` to prevent hanging operations.
* - Use `onAbort`, `onError`, `onLastError`, `onRetry`, `onSettled`, `onSuccess`, `onExecutionTimeout`, and `onTotalExecutionTimeout` for custom side effects.
* - Call `abort()` to cancel ongoing execution and pending retries.
* - Call `reset()` to reset state and cancel execution.
* - Use `getAbortSignal()` to make your async function cancellable.
* - Use dynamic options (functions) for `maxAttempts`, `baseWait`, and `enabled` based on retryer state.
*
* **Important:** This class is designed for single-use execution. Calling `execute()` multiple times
* on the same instance will abort previous executions. For multiple calls, create a new instance
* each time.
*
* @example
* ```typescript
* // Retry a fetch operation up to 5 times with exponential backoff, jitter, and timeouts
* const retryer = new AsyncRetryer(async (url: string) => {
*   const signal = retryer.getAbortSignal()
*   return await fetch(url, { signal })
* }, {
*   maxAttempts: 5,
*   backoff: 'exponential',
*   baseWait: 1000,
*   jitter: 0.1, // Add 10% random variation to prevent thundering herd
*   maxExecutionTime: 5000, // Abort individual calls after 5 seconds
*   maxTotalExecutionTime: 30000, // Abort entire operation after 30 seconds
*   onRetry: (attempt, error) => console.log(`Retry attempt ${attempt} after error:`, error),
*   onSuccess: (result) => console.log('Success:', result),
*   onError: (error) => console.error('Error:', error),
*   onLastError: (error) => console.error('All retries failed:', error),
* })
*
* const result = await retryer.execute('/api/data')
* ```
*
* @template TFn The async function type to be retried.
*/
var AsyncRetryer = class {
	#abortController = null;
	/**
	* Creates a new AsyncRetryer instance
	* @param fn The async function to retry
	* @param initialOptions Configuration options for the retryer
	*/
	constructor(fn, initialOptions = {}) {
		this.fn = fn;
		this.store = new _tanstack_store.Store(getDefaultAsyncRetryerState());
		this.setOptions = (newOptions) => {
			this.options = {
				...this.options,
				...newOptions
			};
		};
		this.execute = async (...args) => {
			if (!this.#getEnabled()) return;
			this.abort("new-execution");
			const startTime = Date.now();
			let lastError;
			let result;
			this.#abortController = new AbortController();
			const signal = this.#abortController.signal;
			this.#setState({
				isExecuting: true,
				currentAttempt: 0,
				lastError: void 0
			});
			let totalTimeoutId;
			if (this.options.maxTotalExecutionTime !== Infinity) totalTimeoutId = setTimeout(() => {
				this.options.onTotalExecutionTimeout?.(this);
				this.abort("total-timeout");
			}, this.options.maxTotalExecutionTime);
			let isLastAttempt = false;
			for (let attempt = 1; attempt <= this.#getMaxAttempts(); attempt++) {
				isLastAttempt = attempt === this.#getMaxAttempts();
				this.#setState({ currentAttempt: attempt });
				try {
					if (signal.aborted) return;
					const currentTotalTime = Date.now() - startTime;
					if (this.options.maxTotalExecutionTime !== Infinity && currentTotalTime >= this.options.maxTotalExecutionTime) {
						this.options.onTotalExecutionTimeout?.(this);
						this.abort("total-timeout");
						return;
					}
					if (this.options.maxExecutionTime === Infinity) result = await this.fn(...args);
					else result = await Promise.race([this.fn(...args), new Promise((_, reject) => {
						const timeout = setTimeout(() => {
							this.options.onExecutionTimeout?.(this);
							this.abort("execution-timeout");
							reject(/* @__PURE__ */ new Error(`Execution timeout: ${this.options.maxExecutionTime}ms exceeded`));
						}, this.options.maxExecutionTime);
						signal.addEventListener("abort", () => {
							clearTimeout(timeout);
							reject(/* @__PURE__ */ new Error("Aborted"));
						}, { once: true });
					})]);
					if (signal.aborted) return;
					const totalTime = Date.now() - startTime;
					this.#setState({
						executionCount: this.store.state.executionCount + 1,
						isExecuting: false,
						lastExecutionTime: Date.now(),
						totalExecutionTime: totalTime,
						currentAttempt: 0,
						lastResult: result
					});
					this.options.onSuccess?.(result, args, this);
					return result;
				} catch (error) {
					if (error && typeof error === "object" && "name" in error && error.name === "AbortError") return;
					lastError = error instanceof Error ? error : new Error(String(error));
					this.#setState({ lastError });
					this.options.onError?.(lastError, args, this);
					if (attempt < this.#getMaxAttempts()) {
						this.options.onRetry?.(attempt, lastError, this);
						const wait = this.#calculateWait(attempt);
						if (wait > 0) {
							this.#setState({
								isExecuting: true,
								currentAttempt: attempt + 1
							});
							await new Promise((resolve) => {
								const timeout = setTimeout(() => {
									signal.removeEventListener("abort", onAbort);
									resolve();
								}, wait);
								const onAbort = () => {
									clearTimeout(timeout);
									signal.removeEventListener("abort", onAbort);
									resolve();
								};
								signal.addEventListener("abort", onAbort);
							});
							if (signal.aborted) return;
						}
					}
				} finally {
					this.options.onSettled?.(args, this);
				}
			}
			if (totalTimeoutId) clearTimeout(totalTimeoutId);
			this.#setState({ isExecuting: false });
			this.options.onLastError?.(lastError, this);
			this.options.onSettled?.(args, this);
			if (this.options.throwOnError === "last" && isLastAttempt || this.options.throwOnError === true) throw lastError;
		};
		this.getAbortSignal = () => {
			return this.#abortController?.signal ?? null;
		};
		this.abort = (reason = "manual") => {
			if (this.#abortController) {
				this.#abortController.abort();
				this.#abortController = null;
				this.#setState({ isExecuting: false });
				this.options.onAbort?.(reason, this);
			}
		};
		this.reset = () => {
			this.#setState(getDefaultAsyncRetryerState());
		};
		this.key = initialOptions.key;
		this.options = {
			...defaultOptions,
			...initialOptions,
			throwOnError: initialOptions.throwOnError ?? (initialOptions.onError ? false : defaultOptions.throwOnError)
		};
		this.#setState(this.options.initialState ?? {});
		if (this.key) require_event_client.pacerEventClient.on("d-AsyncRetryer", (event) => {
			if (event.payload.key !== this.key) return;
			this.#setState(event.payload.store.state);
			this.setOptions(event.payload.options);
		});
	}
	#setState = (newState) => {
		this.store.setState((state) => {
			const combinedState = {
				...state,
				...newState
			};
			const { isExecuting, currentAttempt } = combinedState;
			return {
				...combinedState,
				status: !this.#getEnabled() ? "disabled" : isExecuting && currentAttempt === 1 ? "executing" : isExecuting && currentAttempt > 1 ? "retrying" : "idle"
			};
		});
		require_event_client.emitChange("AsyncRetryer", this);
	};
	#getEnabled = () => {
		return !!require_utils.parseFunctionOrValue(this.options.enabled, this);
	};
	#getMaxAttempts = () => {
		return require_utils.parseFunctionOrValue(this.options.maxAttempts, this);
	};
	#getBaseWait = () => {
		return require_utils.parseFunctionOrValue(this.options.baseWait, this);
	};
	#getMaxWait = () => {
		return require_utils.parseFunctionOrValue(this.options.maxWait, this);
	};
	#calculateJitter = (waitTime) => {
		const jitterAmount = this.options.jitter;
		if (jitterAmount <= 0) return 0;
		try {
			const crypto = typeof globalThis !== "undefined" ? globalThis.crypto : void 0;
			if (crypto?.getRandomValues) {
				const array = new Uint32Array(1);
				crypto.getRandomValues(array);
				const randomFactor = array[0] / 4294967295 * 2 - 1;
				return Math.floor(waitTime * jitterAmount * randomFactor);
			}
		} catch {}
		return 0;
	};
	#calculateWait = (attempt) => {
		const baseWait = this.#getBaseWait();
		let waitTime;
		switch (this.options.backoff) {
			case "linear":
				waitTime = baseWait * attempt;
				break;
			case "exponential":
				waitTime = baseWait * Math.pow(2, attempt - 1);
				break;
			default:
				waitTime = baseWait;
				break;
		}
		waitTime = Math.min(waitTime, this.#getMaxWait());
		const jitter = this.#calculateJitter(waitTime);
		return Math.max(0, waitTime + jitter);
	};
};
/**
* Creates a retry-enabled version of an async function. This is a convenience wrapper
* around the AsyncRetryer class that returns the execute method.
*
* @param fn The async function to add retry functionality to
* @param initialOptions Configuration options for the retry behavior
* @returns A new function that executes the original with retry logic
*
* @example
* ```typescript
* // Define your async function normally
* async function fetchData(url: string) {
*   const response = await fetch(url)
*   if (!response.ok) throw new Error('Request failed')
*   return response.json()
* }
*
* // Create retry-enabled function
* const fetchWithRetry = asyncRetry(fetchData, {
*   maxAttempts: 3,
*   backoff: 'exponential',
*   baseWait: 1000,
*   jitter: 0.1
* })
*
* // Call it multiple times
* const data1 = await fetchWithRetry('/api/data1')
* const data2 = await fetchWithRetry('/api/data2')
* ```
*/
function asyncRetry(fn, initialOptions = {}) {
	return new AsyncRetryer(fn, initialOptions).execute;
}

//#endregion
exports.AsyncRetryer = AsyncRetryer;
exports.asyncRetry = asyncRetry;
exports.asyncRetryerOptions = asyncRetryerOptions;
