// @tanstack/pacer/dist/async-throttler.js
import { parseFunctionOrValue } from "./utils.js";
import { emitChange, pacerEventClient } from "./event-client.js";
import { AsyncRetryer } from "./async-retryer.js";
import { Store } from "@tanstack/store";

//#region src/async-throttler.ts
function getDefaultAsyncThrottlerState() {
	return {
		errorCount: 0,
		isExecuting: false,
		isPending: false,
		lastArgs: void 0,
		lastExecutionTime: 0,
		lastResult: void 0,
		maybeExecuteCount: 0,
		nextExecutionTime: void 0,
		settleCount: 0,
		status: "idle",
		successCount: 0
	};
}
/**
* Utility function for sharing common `AsyncThrottlerOptions` options between different `AsyncThrottler` instances.
*/
function asyncThrottlerOptions(options) {
	return options;
}
const defaultOptions = {
	asyncRetryerOptions: { maxAttempts: 1 },
	enabled: true,
	leading: true,
	trailing: true,
	wait: 0
};
/**
* A class that creates an async throttled function.
*
* Async vs Sync Versions:
* The async version provides advanced features over the sync Throttler:
* - Returns promises that can be awaited for throttled function results
* - Built-in retry support via AsyncRetryer integration
* - Abort support to cancel in-flight executions
* - Cancel support to prevent pending executions from starting
* - Comprehensive error handling with onError callbacks and throwOnError control
* - Detailed execution tracking (success/error/settle counts)
* - Waits for ongoing executions to complete before scheduling the next one
*
* The sync Throttler is lighter weight and simpler when you don't need async features,
* return values, or execution control.
*
* What is Throttling?
* Throttling limits how often a function can be executed, allowing only one execution within a specified time window.
* Unlike debouncing which resets the delay timer on each call, throttling ensures the function executes at a
* regular interval regardless of how often it's called.
*
* This is useful for rate-limiting API calls, handling scroll/resize events, or any scenario where you want to
* ensure a maximum execution frequency.
*
* Error Handling:
* - If an `onError` handler is provided, it will be called with the error and throttler instance
* - If `throwOnError` is true (default when no onError handler is provided), the error will be thrown
* - If `throwOnError` is false (default when onError handler is provided), the error will be swallowed
* - Both onError and throwOnError can be used together - the handler will be called before any error is thrown
* - The error state can be checked using the underlying AsyncThrottler instance
*
* State Management:
* - Uses TanStack Store for reactive state management
* - Use `initialState` to provide initial state values when creating the async throttler
* - Use `onSuccess` callback to react to successful function execution and implement custom logic
* - Use `onError` callback to react to function execution errors and implement custom error handling
* - Use `onSettled` callback to react to function execution completion (success or error) and implement custom logic
* - The state includes error count, execution status, last execution time, and success/settle counts
* - State can be accessed via `asyncThrottler.store.state` when using the class directly
* - When using framework adapters (React/Solid), state is accessed from `asyncThrottler.state`
*
* @example
* ```ts
* const throttler = new AsyncThrottler(async (value: string) => {
*   const result = await saveToAPI(value);
*   return result; // Return value is preserved
* }, {
*   wait: 1000,
*   onError: (error) => {
*     console.error('API call failed:', error);
*   }
* });
*
* // Will only execute once per second no matter how often called
* // Returns the API response directly
* const result = await throttler.maybeExecute(inputElement.value);
* ```
*/
var AsyncThrottler = class {
	#timeoutId = null;
	#resolvePreviousPromise = null;
	constructor(fn, initialOptions) {
		this.fn = fn;
		this.store = new Store(getDefaultAsyncThrottlerState());
		this.asyncRetryers = /* @__PURE__ */ new Map();
		this.setOptions = (newOptions) => {
			this.options = {
				...this.options,
				...newOptions
			};
			if (!this.#getEnabled()) this.cancel();
		};
		this.maybeExecute = async (...args) => {
			if (!this.#getEnabled()) return void 0;
			this.#resolvePreviousPromiseInternal();
			this.#setState({
				maybeExecuteCount: this.store.state.maybeExecuteCount + 1,
				lastArgs: args
			});
			const wait = this.#getWait();
			const thisMaybeExecuteNumber = this.store.state.maybeExecuteCount;
			for (let maxNumIterations = wait / 10; this.store.state.isExecuting && maxNumIterations > 0; maxNumIterations--) {
				await new Promise((resolve) => setTimeout(resolve, 10));
				if (this.store.state.maybeExecuteCount !== thisMaybeExecuteNumber) return this.store.state.lastResult;
			}
			const now = Date.now();
			const timeSinceLastExecution = now - this.store.state.lastExecutionTime;
			if (this.options.leading && !this.store.state.isPending && timeSinceLastExecution >= wait) await this.#execute(...args);
			else if (this.options.trailing) {
				this.cancel();
				this.#setState({ isPending: true });
				return new Promise((resolve, reject) => {
					this.#resolvePreviousPromise = resolve;
					const newTimeSinceLastExecution = this.store.state.lastExecutionTime ? now - this.store.state.lastExecutionTime : 0;
					const timeoutDuration = Math.max(0, wait - newTimeSinceLastExecution);
					this.#timeoutId = setTimeout(async () => {
						this.#clearTimeout();
						if (this.store.state.lastArgs !== void 0) try {
							await this.#execute(...this.store.state.lastArgs);
						} catch (error) {
							reject(error);
						}
						this.#resolvePreviousPromise = null;
						resolve(this.store.state.lastResult);
					}, timeoutDuration);
				});
			}
			return this.store.state.lastResult;
		};
		this.flush = async () => {
			if (this.store.state.isPending && this.store.state.lastArgs) {
				const resolvePromise = this.#resolvePreviousPromise;
				this.#clearTimeout();
				this.#setState({ isPending: false });
				const result = await this.#execute(...this.store.state.lastArgs);
				if (resolvePromise) resolvePromise(result);
				return result;
			}
		};
		this.getAbortSignal = (maybeExecuteCount) => {
			const count = maybeExecuteCount ?? this.store.state.maybeExecuteCount;
			return this.asyncRetryers.get(count)?.getAbortSignal() ?? null;
		};
		this.abort = () => {
			this.asyncRetryers.forEach((retryer) => retryer.abort());
			this.asyncRetryers.clear();
			this.#setState({ isExecuting: false });
		};
		this.cancel = () => {
			this.#clearTimeout();
			if (this.#resolvePreviousPromise) {
				this.#resolvePreviousPromiseInternal();
				this.#resolvePreviousPromise = null;
			}
			this.#setState({ isPending: false });
		};
		this.reset = () => {
			this.#setState(getDefaultAsyncThrottlerState());
			this.asyncRetryers.forEach((retryer) => retryer.reset());
		};
		this.key = initialOptions.key;
		this.options = {
			...defaultOptions,
			...initialOptions,
			throwOnError: initialOptions.throwOnError ?? !initialOptions.onError
		};
		this.#setState(this.options.initialState ?? {});
		if (this.key) pacerEventClient.on("d-AsyncThrottler", (event) => {
			if (event.payload.key !== this.key) return;
			this.#setState(event.payload.store.state);
			this.setOptions(event.payload.options);
		});
	}
	#setState = (newState) => {
		this.store.setState((state) => {
			const combinedState = {
				...state,
				...newState
			};
			const { isPending, isExecuting, settleCount } = combinedState;
			return {
				...combinedState,
				status: !this.#getEnabled() ? "disabled" : isPending ? "pending" : isExecuting ? "executing" : settleCount > 0 ? "settled" : "idle"
			};
		});
		emitChange("AsyncThrottler", this);
	};
	/**
	* Returns the current enabled state of the async throttler
	*/
	#getEnabled = () => {
		return !!parseFunctionOrValue(this.options.enabled, this);
	};
	/**
	* Returns the current wait time in milliseconds
	*/
	#getWait = () => {
		return parseFunctionOrValue(this.options.wait, this);
	};
	#execute = async (...args) => {
		if (!this.#getEnabled()) return void 0;
		const currentMaybeExecute = this.store.state.maybeExecuteCount;
		try {
			this.#setState({ isExecuting: true });
			const currentAsyncRetryer = new AsyncRetryer(this.fn, {
				...this.options.asyncRetryerOptions,
				key: `${this.key}-retryer-${currentMaybeExecute}`
			});
			this.asyncRetryers.set(currentMaybeExecute, currentAsyncRetryer);
			const result = await currentAsyncRetryer.execute(...args);
			this.#setState({
				lastResult: result,
				successCount: this.store.state.successCount + 1
			});
			this.options.onSuccess?.(result, args, this);
		} catch (error) {
			this.#setState({ errorCount: this.store.state.errorCount + 1 });
			this.options.onError?.(error, args, this);
			if (this.options.throwOnError) throw error;
		} finally {
			this.asyncRetryers.delete(currentMaybeExecute);
			const lastExecutionTime = Date.now();
			const wait = this.#getWait();
			const nextExecutionTime = lastExecutionTime + wait;
			this.#setState({
				isExecuting: false,
				isPending: !!this.#timeoutId,
				settleCount: this.store.state.settleCount + 1,
				lastExecutionTime,
				nextExecutionTime
			});
			this.options.onSettled?.(args, this);
			setTimeout(() => {
				if (!this.store.state.isPending) this.#setState({ nextExecutionTime: void 0 });
			}, wait);
		}
		return this.store.state.lastResult;
	};
	#resolvePreviousPromiseInternal = () => {
		if (this.#resolvePreviousPromise) {
			this.#resolvePreviousPromise(this.store.state.lastResult);
			this.#resolvePreviousPromise = null;
		}
	};
	#clearTimeout = () => {
		if (this.#timeoutId) {
			clearTimeout(this.#timeoutId);
			this.#timeoutId = null;
		}
	};
};
/**
* Creates an async throttled function that limits how often the function can execute.
* The throttled function will execute at most once per wait period, even if called multiple times.
* If called while executing, it will wait until execution completes before scheduling the next call.
*
* Async vs Sync Versions:
* The async version provides advanced features over the sync throttle function:
* - Returns promises that can be awaited for throttled function results
* - Built-in retry support via AsyncRetryer integration
* - Abort support to cancel in-flight executions
* - Cancel support to prevent pending executions from starting
* - Comprehensive error handling with onError callbacks and throwOnError control
* - Detailed execution tracking (success/error/settle counts)
* - Waits for ongoing executions to complete before scheduling the next one
*
* The sync throttle function is lighter weight and simpler when you don't need async features,
* return values, or execution control.
*
* What is Throttling?
* Throttling limits how often a function can be executed, allowing only one execution within a specified time window.
* Unlike debouncing which resets the delay timer on each call, throttling ensures the function executes at a
* regular interval regardless of how often it's called.
*
* Configuration Options:
* - `wait`: Time window in milliseconds during which the function can only execute once (required)
* - `leading`: Execute immediately when called (default: true)
* - `trailing`: Execute on the trailing edge of the wait period (default: true)
* - `enabled`: Whether the throttler is enabled (default: true)
* - `asyncRetryerOptions`: Configure retry behavior for executions
*
* Error Handling:
* - If an `onError` handler is provided, it will be called with the error and throttler instance
* - If `throwOnError` is true (default when no onError handler is provided), the error will be thrown
* - If `throwOnError` is false (default when onError handler is provided), the error will be swallowed
* - Both onError and throwOnError can be used together - the handler will be called before any error is thrown
* - The error state can be checked using the underlying AsyncThrottler instance
*
* State Management:
* - Uses TanStack Store for reactive state management
* - Use `initialState` to provide initial state values when creating the async throttler
* - Use `onSuccess` callback to react to successful function execution and implement custom logic
* - Use `onError` callback to react to function execution errors and implement custom error handling
* - Use `onSettled` callback to react to function execution completion (success or error) and implement custom logic
* - The state includes error count, execution status, last execution time, and success/settle counts
* - State can be accessed via the underlying AsyncThrottler instance's `store.state` property
* - When using framework adapters (React/Solid), state is accessed from the hook's state property
*
* @example
* ```ts
* const throttled = asyncThrottle(async (value: string) => {
*   const result = await saveToAPI(value);
*   return result; // Return value is preserved
* }, {
*   wait: 1000,
*   onError: (error) => {
*     console.error('API call failed:', error);
*   }
* });
*
* // This will execute at most once per second
* // Returns the API response directly
* const result = await throttled(inputElement.value);
* ```
*/
function asyncThrottle(fn, initialOptions) {
	return new AsyncThrottler(fn, initialOptions).maybeExecute;
}

//#endregion
export { AsyncThrottler, asyncThrottle, asyncThrottlerOptions };
