// @tanstack/pacer/dist/async-debouncer.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_utils = require('./utils.cjs');
const require_event_client = require('./event-client.cjs');
const require_async_retryer = require('./async-retryer.cjs');
let _tanstack_store = require("@tanstack/store");

//#region src/async-debouncer.ts
function getDefaultAsyncDebouncerState() {
	return {
		canLeadingExecute: true,
		errorCount: 0,
		isExecuting: false,
		isPending: false,
		lastArgs: void 0,
		lastResult: void 0,
		maybeExecuteCount: 0,
		settleCount: 0,
		status: "idle",
		successCount: 0
	};
}
/**
* Utility function for sharing common `AsyncDebouncerOptions` options between different `AsyncDebouncer` instances.
*/
function asyncDebouncerOptions(options) {
	return options;
}
const defaultOptions = {
	asyncRetryerOptions: { maxAttempts: 1 },
	enabled: true,
	leading: false,
	trailing: true,
	wait: 0
};
/**
* A class that creates an async debounced function.
*
* Async vs Sync Versions:
* The async version provides advanced features over the sync Debouncer:
* - Returns promises that can be awaited for debounced function results
* - Built-in retry support via AsyncRetryer integration
* - Abort support to cancel in-flight executions
* - Cancel support to prevent pending executions from starting
* - Comprehensive error handling with onError callbacks and throwOnError control
* - Detailed execution tracking (success/error/settle counts)
*
* The sync Debouncer is lighter weight and simpler when you don't need async features,
* return values, or execution control.
*
* What is Debouncing?
* Debouncing ensures that a function is only executed after a specified delay has passed since its last invocation.
* Each new invocation resets the delay timer. This is useful for handling frequent events like window resizing
* or input changes where you only want to execute the handler after the events have stopped occurring.
*
* Unlike throttling which allows execution at regular intervals, debouncing prevents any execution until
* the function stops being called for the specified delay period.
*
* Error Handling:
* - If an `onError` handler is provided, it will be called with the error and debouncer instance
* - If `throwOnError` is true (default when no onError handler is provided), the error will be thrown
* - If `throwOnError` is false (default when onError handler is provided), the error will be swallowed
* - Both onError and throwOnError can be used together - the handler will be called before any error is thrown
* - The error state can be checked using the underlying store
*
* State Management:
* - The debouncer uses a reactive store for state management
* - Use `initialState` to provide initial state values when creating the async debouncer
* - The state includes canLeadingExecute, error count, execution status, and success/settle counts
* - State can be accessed via the `store` property and its `state` getter
* - The store is reactive and will notify subscribers of state changes
*
* @example
* ```ts
* const asyncDebouncer = new AsyncDebouncer(async (value: string) => {
*   const results = await searchAPI(value);
*   return results; // Return value is preserved
* }, {
*   wait: 500,
*   onError: (error) => {
*     console.error('Search failed:', error);
*   }
* });
*
* // Called on each keystroke but only executes after 500ms of no typing
* // Returns the API response directly
* const results = await asyncDebouncer.maybeExecute(inputElement.value);
* ```
*/
var AsyncDebouncer = class {
	#timeoutId = null;
	#resolvePreviousPromise = null;
	constructor(fn, initialOptions) {
		this.fn = fn;
		this.store = new _tanstack_store.Store(getDefaultAsyncDebouncerState());
		this.asyncRetryers = /* @__PURE__ */ new Map();
		this.setOptions = (newOptions) => {
			this.options = {
				...this.options,
				...newOptions
			};
			if (!this.#getEnabled()) this.cancel();
		};
		this.maybeExecute = async (...args) => {
			if (!this.#getEnabled()) return void 0;
			this.#cancelPendingExecution();
			this.#setState({
				lastArgs: args,
				maybeExecuteCount: this.store.state.maybeExecuteCount + 1
			});
			if (this.options.leading && this.store.state.canLeadingExecute) {
				this.#setState({ canLeadingExecute: false });
				await this.#execute(...args);
				return this.store.state.lastResult;
			}
			if (this.options.trailing && this.#getEnabled()) this.#setState({ isPending: true });
			return new Promise((resolve, reject) => {
				this.#resolvePreviousPromise = resolve;
				this.#timeoutId = setTimeout(async () => {
					if (this.options.trailing && this.store.state.lastArgs) try {
						await this.#execute(...this.store.state.lastArgs);
					} catch (error) {
						reject(error);
					}
					this.#setState({ canLeadingExecute: true });
					this.#resolvePreviousPromise = null;
					resolve(this.store.state.lastResult);
				}, this.#getWait());
			});
		};
		this.flush = async () => {
			if (this.store.state.isPending && this.store.state.lastArgs) {
				const { lastArgs } = this.store.state;
				this.#cancelPendingExecution();
				return await this.#execute(...lastArgs);
			}
		};
		this.getAbortSignal = (maybeExecuteCount) => {
			const count = maybeExecuteCount ?? this.store.state.maybeExecuteCount;
			return this.asyncRetryers.get(count)?.getAbortSignal() ?? null;
		};
		this.abort = () => {
			this.asyncRetryers.forEach((retryer) => retryer.abort());
			this.asyncRetryers.clear();
			this.#setState({ isExecuting: false });
		};
		this.cancel = () => {
			this.#cancelPendingExecution();
			this.#setState({ canLeadingExecute: true });
		};
		this.reset = () => {
			this.#setState(getDefaultAsyncDebouncerState());
			this.asyncRetryers.forEach((retryer) => retryer.reset());
		};
		this.key = initialOptions.key;
		this.options = {
			...defaultOptions,
			...initialOptions,
			throwOnError: initialOptions.throwOnError ?? !initialOptions.onError
		};
		this.#setState(this.options.initialState ?? {});
		if (this.key) require_event_client.pacerEventClient.on("d-AsyncDebouncer", (event) => {
			if (event.payload.key !== this.key) return;
			this.#setState(event.payload.store.state);
			this.setOptions(event.payload.options);
		});
	}
	#setState = (newState) => {
		this.store.setState((state) => {
			const combinedState = {
				...state,
				...newState
			};
			const { isPending, isExecuting, settleCount } = combinedState;
			return {
				...combinedState,
				status: !this.#getEnabled() ? "disabled" : isPending ? "pending" : isExecuting ? "executing" : settleCount > 0 ? "settled" : "idle"
			};
		});
		require_event_client.emitChange("AsyncDebouncer", this);
	};
	/**
	* Returns the current debouncer enabled state
	*/
	#getEnabled = () => {
		return !!require_utils.parseFunctionOrValue(this.options.enabled, this);
	};
	/**
	* Returns the current debouncer wait state
	*/
	#getWait = () => {
		return require_utils.parseFunctionOrValue(this.options.wait, this);
	};
	#execute = async (...args) => {
		if (!this.#getEnabled()) return void 0;
		const currentMaybeExecuteCount = this.store.state.maybeExecuteCount + 1;
		try {
			this.#setState({ isExecuting: true });
			const currentAsyncRetryer = new require_async_retryer.AsyncRetryer(this.fn, {
				...this.options.asyncRetryerOptions,
				key: `${this.key}-retryer-${currentMaybeExecuteCount}`
			});
			this.asyncRetryers.set(currentMaybeExecuteCount, currentAsyncRetryer);
			const result = await currentAsyncRetryer.execute(...args);
			this.#setState({
				lastResult: result,
				successCount: this.store.state.successCount + 1
			});
			this.options.onSuccess?.(result, args, this);
		} catch (error) {
			this.#setState({ errorCount: this.store.state.errorCount + 1 });
			this.options.onError?.(error, args, this);
			if (this.options.throwOnError) throw error;
		} finally {
			this.asyncRetryers.delete(currentMaybeExecuteCount);
			this.#setState({
				isExecuting: false,
				isPending: false,
				lastArgs: void 0,
				settleCount: this.store.state.settleCount + 1
			});
			this.options.onSettled?.(args, this);
		}
		return this.store.state.lastResult;
	};
	#resolvePreviousPromiseInternal = () => {
		if (this.#resolvePreviousPromise) {
			this.#resolvePreviousPromise(this.store.state.lastResult);
			this.#resolvePreviousPromise = null;
		}
	};
	#clearTimeout = () => {
		if (this.#timeoutId) {
			clearTimeout(this.#timeoutId);
			this.#timeoutId = null;
		}
	};
	/**
	* Internal cancel without resetting the leading execute state
	*/
	#cancelPendingExecution = () => {
		this.#clearTimeout();
		this.#resolvePreviousPromiseInternal();
		this.#setState({
			isPending: false,
			lastArgs: void 0
		});
	};
};
/**
* Creates an async debounced function that delays execution until after a specified wait time.
* The debounced function will only execute once the wait period has elapsed without any new calls.
* If called again during the wait period, the timer resets and a new wait period begins.
*
* Async vs Sync Versions:
* The async version provides advanced features over the sync debounce function:
* - Returns promises that can be awaited for debounced function results
* - Built-in retry support via AsyncRetryer integration
* - Abort support to cancel in-flight executions
* - Cancel support to prevent pending executions from starting
* - Comprehensive error handling with onError callbacks and throwOnError control
* - Detailed execution tracking (success/error/settle counts)
*
* The sync debounce function is lighter weight and simpler when you don't need async features,
* return values, or execution control.
*
* What is Debouncing?
* Debouncing ensures that a function is only executed after a specified delay has passed since its last invocation.
* Each new invocation resets the delay timer. This is useful for handling frequent events like window resizing
* or input changes where you only want to execute the handler after the events have stopped occurring.
*
* Configuration Options:
* - `wait`: Delay in milliseconds to wait after the last call (required)
* - `leading`: Execute on the leading edge of the timeout (default: false)
* - `trailing`: Execute on the trailing edge of the timeout (default: true)
* - `enabled`: Whether the debouncer is enabled (default: true)
* - `asyncRetryerOptions`: Configure retry behavior for executions
*
* Error Handling:
* - If an `onError` handler is provided, it will be called with the error and debouncer instance
* - If `throwOnError` is true (default when no onError handler is provided), the error will be thrown
* - If `throwOnError` is false (default when onError handler is provided), the error will be swallowed
* - The error state can be checked using the underlying AsyncDebouncer instance
* - Both onError and throwOnError can be used together - the handler will be called before any error is thrown
*
* State Management:
* - Uses TanStack Store for reactive state management
* - Use `initialState` to provide initial state values when creating the async debouncer
* - Use `onSuccess` callback to react to successful function execution and implement custom logic
* - Use `onError` callback to react to function execution errors and implement custom error handling
* - Use `onSettled` callback to react to function execution completion (success or error) and implement custom logic
* - The state includes canLeadingExecute, error count, execution status, and success/settle counts
* - State can be accessed via `asyncDebouncer.store.state` when using the class directly
* - When using framework adapters (React/Solid), state is accessed from `asyncDebouncer.state`
*
* @example
* ```ts
* const debounced = asyncDebounce(async (value: string) => {
*   const result = await saveToAPI(value);
*   return result; // Return value is preserved
* }, {
*   wait: 1000,
*   onError: (error) => {
*     console.error('API call failed:', error);
*   },
*   throwOnError: true // Will both log the error and throw it
* });
*
* // Will only execute once, 1 second after the last call
* // Returns the API response directly
* const result = await debounced("third");
* ```
*/
function asyncDebounce(fn, initialOptions) {
	return new AsyncDebouncer(fn, initialOptions).maybeExecute;
}

//#endregion
exports.AsyncDebouncer = AsyncDebouncer;
exports.asyncDebounce = asyncDebounce;
exports.asyncDebouncerOptions = asyncDebouncerOptions;
