// @tanstack/pacer/dist/async-batcher.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_utils = require('./utils.cjs');
const require_event_client = require('./event-client.cjs');
const require_async_retryer = require('./async-retryer.cjs');
let _tanstack_store = require("@tanstack/store");

//#region src/async-batcher.ts
function getDefaultAsyncBatcherState() {
	return {
		errorCount: 0,
		executeCount: 0,
		failedItems: [],
		isEmpty: true,
		isExecuting: false,
		isPending: false,
		items: [],
		lastResult: void 0,
		settleCount: 0,
		size: 0,
		status: "idle",
		successCount: 0,
		totalItemsProcessed: 0,
		totalItemsFailed: 0
	};
}
/**
* Utility function for sharing common `AsyncBatcherOptions` options between different `AsyncBatcher` instances.
*
*/
function asyncBatcherOptions(options) {
	return options;
}
const defaultOptions = {
	asyncRetryerOptions: { maxAttempts: 1 },
	getShouldExecute: () => false,
	maxSize: Infinity,
	started: true,
	throwOnError: true,
	wait: Infinity
};
/**
* A class that collects items and processes them in batches asynchronously.
*
* Async vs Sync Versions:
* The async version provides advanced features over the sync Batcher:
* - Returns promises that can be awaited for batch results
* - Built-in retry support via AsyncRetryer integration
* - Abort support to cancel in-flight batch executions
* - Cancel support to prevent pending batches from starting
* - Comprehensive error handling with onError callbacks and throwOnError control
* - Detailed execution tracking (success/error/settle counts)
*
* The sync Batcher is lighter weight and simpler when you don't need async features,
* return values, or execution control.
*
* What is Batching?
* Batching is a technique for grouping multiple operations together to be processed as a single unit.
*
* The AsyncBatcher provides a flexible way to implement async batching with configurable:
* - Maximum batch size (number of items per batch)
* - Time-based batching (process after X milliseconds)
* - Custom batch processing logic via getShouldExecute
* - Event callbacks for monitoring batch operations
* - Error handling for failed batch operations
*
* Error Handling:
* - If an `onError` handler is provided, it will be called with the error, the batch of items that failed, and batcher instance
* - If `throwOnError` is true (default when no onError handler is provided), the error will be thrown
* - If `throwOnError` is false (default when onError handler is provided), the error will be swallowed
* - Both onError and throwOnError can be used together - the handler will be called before any error is thrown
* - The error state can be checked using the AsyncBatcher instance
*
* State Management:
* - Uses TanStack Store for reactive state management
* - Use `initialState` to provide initial state values when creating the async batcher
* - Use `onSuccess` callback to react to successful batch execution and implement custom logic
* - Use `onError` callback to react to batch execution errors and implement custom error handling
* - Use `onSettled` callback to react to batch execution completion (success or error) and implement custom logic
* - Use `onExecute` callback to react to batch execution and implement custom logic
* - Use `onItemsChange` callback to react to items being added or removed from the batcher
* - The state includes total items processed, success/error counts, and execution status
* - State can be accessed via `asyncBatcher.store.state` when using the class directly
* - When using framework adapters (React/Solid), state is accessed from `asyncBatcher.state`
*
* @example
* ```ts
* const batcher = new AsyncBatcher<number>(
*   async (items) => {
*     const result = await processItems(items);
*     console.log('Processing batch:', items);
*     return result;
*   },
*   {
*     maxSize: 5,
*     wait: 2000,
*     onSuccess: (result) => console.log('Batch succeeded:', result),
*     onError: (error) => console.error('Batch failed:', error)
*   }
* );
*
* batcher.addItem(1);
* batcher.addItem(2);
* // After 2 seconds or when 5 items are added, whichever comes first,
* // the batch will be processed and the result will be available
* // batcher.execute() // manually trigger a batch
* ```
*/
var AsyncBatcher = class {
	#timeoutId = null;
	constructor(fn, initialOptions) {
		this.fn = fn;
		this.store = new _tanstack_store.Store(getDefaultAsyncBatcherState());
		this.asyncRetryers = /* @__PURE__ */ new Map();
		this.setOptions = (newOptions) => {
			this.options = {
				...this.options,
				...newOptions
			};
		};
		this.addItem = async (item) => {
			this.#setState({
				items: [...this.store.state.items, item],
				isPending: this.options.wait !== Infinity
			});
			this.options.onItemsChange?.(this);
			if (this.store.state.items.length >= this.options.maxSize || this.options.getShouldExecute(this.store.state.items, this)) return await this.#execute();
			else if (this.options.wait !== Infinity) {
				this.#clearTimeout();
				this.#timeoutId = setTimeout(() => this.#execute(), this.#getWait());
				await new Promise((resolve) => setTimeout(resolve, this.#getWait()));
			}
		};
		this.flush = async () => {
			this.#clearTimeout();
			return await this.#execute();
		};
		this.peekAllItems = () => {
			return [...this.store.state.items];
		};
		this.peekFailedItems = () => {
			return [...this.store.state.failedItems];
		};
		this.clear = () => {
			this.#setState({
				items: [],
				failedItems: [],
				isPending: false
			});
		};
		this.getAbortSignal = (executeCount) => {
			const count = executeCount ?? this.store.state.executeCount;
			return this.asyncRetryers.get(count)?.getAbortSignal() ?? null;
		};
		this.abort = () => {
			this.asyncRetryers.forEach((retryer) => retryer.abort());
			this.asyncRetryers.clear();
			this.#setState({ isExecuting: false });
		};
		this.cancel = () => {
			this.#clearTimeout();
			this.#setState({ isPending: false });
		};
		this.reset = () => {
			this.#setState(getDefaultAsyncBatcherState());
			this.options.onItemsChange?.(this);
			this.asyncRetryers.forEach((retryer) => retryer.reset());
		};
		this.key = initialOptions.key;
		this.options = {
			...defaultOptions,
			...initialOptions,
			throwOnError: initialOptions.throwOnError ?? !initialOptions.onError
		};
		this.#setState(this.options.initialState ?? {});
		if (this.key) require_event_client.pacerEventClient.on("d-AsyncBatcher", (event) => {
			if (event.payload.key !== this.key) return;
			this.#setState(event.payload.store.state);
			this.setOptions(event.payload.options);
		});
	}
	#setState = (newState) => {
		this.store.setState((state) => {
			const combinedState = {
				...state,
				...newState
			};
			const { isExecuting, isPending, items } = combinedState;
			const size = items.length;
			const isEmpty = size === 0;
			return {
				...combinedState,
				isEmpty,
				size,
				status: isExecuting ? "executing" : isPending ? "pending" : isEmpty ? "idle" : "populated"
			};
		});
		require_event_client.emitChange("AsyncBatcher", this);
	};
	#getWait = () => {
		return require_utils.parseFunctionOrValue(this.options.wait, this);
	};
	/**
	* Processes the current batch of items asynchronously.
	* This method will automatically be triggered if the batcher is running and any of these conditions are met:
	* - The number of items reaches maxSize
	* - The wait duration has elapsed
	* - The getShouldExecute function returns true upon adding an item
	*
	* You can also call this method manually to process the current batch at any time.
	*
	* @returns A promise that resolves with the result of the batch function, or undefined if an error occurred and was handled by onError
	* @throws The error from the batch function if no onError handler is configured or throwOnError is true
	*/
	#execute = async () => {
		if (this.store.state.items.length === 0) return;
		const currentExecuteCount = this.store.state.executeCount + 1;
		const batch = this.peekAllItems();
		this.clear();
		this.options.onItemsChange?.(this);
		this.#setState({
			isExecuting: true,
			executeCount: currentExecuteCount
		});
		try {
			const currentAsyncRetryer = new require_async_retryer.AsyncRetryer(this.fn, this.options.asyncRetryerOptions);
			this.asyncRetryers.set(currentExecuteCount, currentAsyncRetryer);
			const result = await currentAsyncRetryer.execute(batch);
			this.#setState({
				totalItemsProcessed: this.store.state.totalItemsProcessed + batch.length,
				lastResult: result,
				successCount: this.store.state.successCount + 1
			});
			this.options.onSuccess?.(result, batch, this);
			return result;
		} catch (error) {
			this.#setState({
				errorCount: this.store.state.errorCount + 1,
				failedItems: [...this.store.state.failedItems, ...batch],
				totalItemsFailed: this.store.state.totalItemsFailed + batch.length
			});
			this.options.onError?.(error, batch, this);
			if (this.options.throwOnError) throw error;
			return;
		} finally {
			this.asyncRetryers.delete(currentExecuteCount);
			this.#setState({
				isExecuting: false,
				settleCount: this.store.state.settleCount + 1
			});
			this.options.onSettled?.(batch, this);
		}
	};
	#clearTimeout = () => {
		if (this.#timeoutId) {
			clearTimeout(this.#timeoutId);
			this.#timeoutId = null;
		}
	};
};
/**
* Creates an async batcher that processes items in batches.
*
* Async vs Sync Versions:
* The async version provides advanced features over the sync batch function:
* - Returns promises that can be awaited for batch results
* - Built-in retry support via AsyncRetryer integration
* - Abort support to cancel in-flight batch executions
* - Cancel support to prevent pending batches from starting
* - Comprehensive error handling with onError callbacks and throwOnError control
* - Detailed execution tracking (success/error/settle counts)
*
* The sync batch function is lighter weight and simpler when you don't need async features,
* return values, or execution control.
*
* What is Batching?
* Batching is a technique for grouping multiple operations together to be processed as a single unit.
*
* Configuration Options:
* - `maxSize`: Maximum number of items per batch (default: Infinity)
* - `wait`: Time to wait before processing batch (default: Infinity)
* - `getShouldExecute`: Custom logic to trigger batch processing
* - `asyncRetryerOptions`: Configure retry behavior for batch executions
* - `started`: Whether to start processing immediately (default: true)
*
* Error Handling:
* - If an `onError` handler is provided, it will be called with the error, the batch of items that failed, and batcher instance
* - If `throwOnError` is true (default when no onError handler is provided), the error will be thrown
* - If `throwOnError` is false (default when onError handler is provided), the error will be swallowed
* - Both onError and throwOnError can be used together - the handler will be called before any error is thrown
* - The error state can be checked using the underlying AsyncBatcher instance
*
* State Management:
* - Uses TanStack Store for reactive state management
* - Use `initialState` to provide initial state values when creating the async batcher
* - Use `onSuccess` callback to react to successful batch execution and implement custom logic
* - Use `onError` callback to react to batch execution errors and implement custom error handling
* - Use `onSettled` callback to react to batch execution completion (success or error) and implement custom logic
* - Use `onItemsChange` callback to react to items being added or removed from the batcher
* - The state includes total items processed, success/error counts, and execution status
* - State can be accessed via the underlying AsyncBatcher instance's `store.state` property
* - When using framework adapters (React/Solid), state is accessed from the hook's state property
*
* @example
* ```ts
* const batchItems = asyncBatch<number>(
*   async (items) => {
*     const result = await processApiCall(items);
*     console.log('Processing:', items);
*     return result;
*   },
*   {
*     maxSize: 3,
*     wait: 1000,
*     onSuccess: (result) => console.log('Batch succeeded:', result),
*     onError: (error) => console.error('Batch failed:', error)
*   }
* );
*
* batchItems(1);
* batchItems(2);
* batchItems(3); // Triggers batch processing
* ```
*/
function asyncBatch(fn, options) {
	return new AsyncBatcher(fn, options).addItem;
}

//#endregion
exports.AsyncBatcher = AsyncBatcher;
exports.asyncBatch = asyncBatch;
exports.asyncBatcherOptions = asyncBatcherOptions;
