// @tanstack/pacer/dist/batcher.js
import { parseFunctionOrValue } from "./utils.js";
import { emitChange, pacerEventClient } from "./event-client.js";
import { Store } from "@tanstack/store";

//#region src/batcher.ts
function getDefaultBatcherState() {
	return {
		executionCount: 0,
		isEmpty: true,
		isPending: false,
		totalItemsProcessed: 0,
		items: [],
		size: 0,
		status: "idle"
	};
}
const defaultOptions = {
	getShouldExecute: () => false,
	maxSize: Infinity,
	started: true,
	wait: Infinity
};
/**
* A class that collects items and processes them in batches.
*
* Batching is a technique for grouping multiple operations together to be processed as a single unit.
* This synchronous version is lighter weight and often all you need - upgrade to AsyncBatcher when you need promises, retry support, abort/cancel capabilities, or advanced error handling.
*
* The Batcher provides a flexible way to implement batching with configurable:
* - Maximum batch size (number of items per batch)
* - Time-based batching (process after X milliseconds)
* - Custom batch processing logic via getShouldExecute
* - Event callbacks for monitoring batch operations
*
* State Management:
* - Uses TanStack Store for reactive state management
* - Use `initialState` to provide initial state values when creating the batcher
* - Use `onExecute` callback to react to batch execution and implement custom logic
* - Use `onItemsChange` callback to react to items being added or removed from the batcher
* - The state includes batch execution count, total items processed, items, and running status
* - State can be accessed via `batcher.store.state` when using the class directly
* - When using framework adapters (React/Solid), state is accessed from `batcher.state`
*
* @example
* ```ts
* const batcher = new Batcher<number>(
*   (items) => console.log('Processing batch:', items),
*   {
*     maxSize: 5,
*     wait: 2000,
*     onExecute: (batch, batcher) => console.log('Batch executed:', batch)
*   }
* );
*
* batcher.addItem(1);
* batcher.addItem(2);
* // After 2 seconds or when 5 items are added, whichever comes first,
* // the batch will be processed
* // batcher.flush() // manually trigger a batch
* ```
*/
var Batcher = class {
	#timeoutId = null;
	constructor(fn, initialOptions) {
		this.fn = fn;
		this.store = new Store(getDefaultBatcherState());
		this.setOptions = (newOptions) => {
			this.options = {
				...this.options,
				...newOptions
			};
		};
		this.addItem = (item) => {
			this.#setState({
				items: [...this.store.state.items, item],
				isPending: this.options.wait !== Infinity
			});
			this.options.onItemsChange?.(this);
			if (this.store.state.items.length >= this.options.maxSize || this.options.getShouldExecute(this.store.state.items, this)) this.#execute();
			else if (this.options.wait !== Infinity) {
				this.#clearTimeout();
				this.#timeoutId = setTimeout(() => this.#execute(), this.#getWait());
			}
		};
		this.flush = () => {
			this.#clearTimeout();
			this.#execute();
		};
		this.peekAllItems = () => {
			return [...this.store.state.items];
		};
		this.clear = () => {
			this.#setState({
				items: [],
				isPending: false
			});
		};
		this.cancel = () => {
			this.#clearTimeout();
			this.#setState({ isPending: false });
		};
		this.reset = () => {
			this.#setState(getDefaultBatcherState());
			this.options.onItemsChange?.(this);
		};
		this.key = initialOptions.key;
		this.options = {
			...defaultOptions,
			...initialOptions
		};
		this.#setState(this.options.initialState ?? {});
		if (this.key) pacerEventClient.on("d-Batcher", (event) => {
			if (event.payload.key !== this.key) return;
			this.#setState(event.payload.store.state);
			this.setOptions(event.payload.options);
		});
	}
	#setState = (newState) => {
		this.store.setState((state) => {
			const combinedState = {
				...state,
				...newState
			};
			const { isPending, items } = combinedState;
			const size = items.length;
			const isEmpty = size === 0;
			return {
				...combinedState,
				isEmpty,
				size,
				status: isPending ? "pending" : "idle"
			};
		});
		emitChange("Batcher", this);
	};
	#getWait = () => {
		return parseFunctionOrValue(this.options.wait, this);
	};
	/**
	* Processes the current batch of items.
	* This method will automatically be triggered if the batcher is running and any of these conditions are met:
	* - The number of items reaches batchSize
	* - The wait duration has elapsed
	* - The getShouldExecute function returns true upon adding an item
	*
	* You can also call this method manually to process the current batch at any time.
	*/
	#execute = () => {
		if (this.store.state.items.length === 0) return;
		const batch = this.peekAllItems();
		this.clear();
		this.options.onItemsChange?.(this);
		this.fn(batch);
		this.#setState({
			executionCount: this.store.state.executionCount + 1,
			totalItemsProcessed: this.store.state.totalItemsProcessed + batch.length
		});
		this.options.onExecute?.(batch, this);
	};
	#clearTimeout = () => {
		if (this.#timeoutId) {
			clearTimeout(this.#timeoutId);
			this.#timeoutId = null;
		}
	};
};
/**
* Creates a batcher that processes items in batches.
*
* This synchronous version is lighter weight and often all you need - upgrade to asyncBatch when you need promises, retry support, abort/cancel capabilities, or advanced error handling.
*
* @example
* ```ts
* const batchItems = batch<number>(
*   (items) => console.log('Processing:', items),
*   {
*     maxSize: 3,
*     onExecute: (batch, batcher) => console.log('Batch executed:', batch)
*   }
* );
*
* batchItems(1);
* batchItems(2);
* batchItems(3); // Triggers batch processing
* ```
*/
function batch(fn, options) {
	return new Batcher(fn, options).addItem;
}

//#endregion
export { Batcher, batch };
