// @tanstack/pacer/dist/debouncer.js
import { parseFunctionOrValue } from "./utils.js";
import { emitChange, pacerEventClient } from "./event-client.js";
import { Store } from "@tanstack/store";

//#region src/debouncer.ts
function getDefaultDebouncerState() {
	return {
		canLeadingExecute: true,
		executionCount: 0,
		isPending: false,
		lastArgs: void 0,
		status: "idle",
		maybeExecuteCount: 0
	};
}
/**
* Utility function for sharing common `DebouncerOptions` options between different `Debouncer` instances.
*/
function debouncerOptions(options) {
	return options;
}
const defaultOptions = {
	enabled: true,
	leading: false,
	trailing: true,
	wait: 0
};
/**
* A class that creates a debounced function.
*
* Debouncing ensures that a function is only executed after a certain amount of time has passed
* since its last invocation. This is useful for handling frequent events like window resizing,
* scroll events, or input changes where you want to limit the rate of execution.
* This synchronous version is lighter weight and often all you need - upgrade to AsyncDebouncer when you need promises, retry support, abort/cancel capabilities, or advanced error handling.
*
* The debounced function can be configured to execute either at the start of the delay period
* (leading edge) or at the end (trailing edge, default). Each new call during the wait period
* will reset the timer.
*
* State Management:
* - Uses TanStack Store for reactive state management
* - Use `initialState` to provide initial state values when creating the debouncer
* - Use `onExecute` callback to react to function execution and implement custom logic
* - The state includes canLeadingExecute, execution count, and isPending status
* - State can be accessed via `debouncer.store.state` when using the class directly
* - When using framework adapters (React/Solid), state is accessed from `debouncer.state`
*
* @example
* ```ts
* const debouncer = new Debouncer((value: string) => {
*   saveToDatabase(value);
* }, { wait: 500 });
*
* // Will only save after 500ms of no new input
* inputElement.addEventListener('input', () => {
*   debouncer.maybeExecute(inputElement.value);
* });
* ```
*/
var Debouncer = class {
	#timeoutId;
	constructor(fn, initialOptions) {
		this.fn = fn;
		this.store = new Store(getDefaultDebouncerState());
		this.setOptions = (newOptions) => {
			this.options = {
				...this.options,
				...newOptions
			};
			if (!this.#getEnabled()) this.cancel();
		};
		this.maybeExecute = (...args) => {
			if (!this.#getEnabled()) return void 0;
			this.#setState({ maybeExecuteCount: this.store.state.maybeExecuteCount + 1 });
			let _didLeadingExecute = false;
			if (this.options.leading && this.store.state.canLeadingExecute) {
				this.#setState({ canLeadingExecute: false });
				_didLeadingExecute = true;
				this.#execute(...args);
			}
			if (this.options.trailing) this.#setState({
				isPending: true,
				lastArgs: args
			});
			if (this.#timeoutId) clearTimeout(this.#timeoutId);
			this.#timeoutId = setTimeout(() => {
				this.#setState({ canLeadingExecute: true });
				if (this.options.trailing && !_didLeadingExecute) this.#execute(...args);
			}, this.#getWait());
		};
		this.flush = () => {
			if (this.store.state.isPending && this.store.state.lastArgs) {
				this.#clearTimeout();
				this.#execute(...this.store.state.lastArgs);
			}
		};
		this.cancel = () => {
			this.#clearTimeout();
			this.#setState({
				canLeadingExecute: true,
				isPending: false
			});
		};
		this.reset = () => {
			this.#setState(getDefaultDebouncerState());
		};
		this.key = initialOptions.key;
		this.options = {
			...defaultOptions,
			...initialOptions
		};
		this.#setState(this.options.initialState ?? {});
		if (this.key) pacerEventClient.on("d-Debouncer", (event) => {
			if (event.payload.key !== this.key) return;
			this.#setState(event.payload.store.state);
			this.setOptions(event.payload.options);
		});
	}
	#setState = (newState) => {
		this.store.setState((state) => {
			const combinedState = {
				...state,
				...newState
			};
			const { isPending } = combinedState;
			return {
				...combinedState,
				status: !this.#getEnabled() ? "disabled" : isPending ? "pending" : "idle"
			};
		});
		emitChange("Debouncer", this);
	};
	/**
	* Returns the current enabled state of the debouncer
	*/
	#getEnabled = () => {
		return !!parseFunctionOrValue(this.options.enabled, this);
	};
	/**
	* Returns the current wait time in milliseconds
	*/
	#getWait = () => {
		return parseFunctionOrValue(this.options.wait, this);
	};
	#execute = (...args) => {
		if (!this.#getEnabled()) return void 0;
		this.fn(...args);
		this.#setState({
			executionCount: this.store.state.executionCount + 1,
			isPending: false,
			lastArgs: void 0
		});
		this.options.onExecute?.(args, this);
	};
	#clearTimeout = () => {
		if (this.#timeoutId) {
			clearTimeout(this.#timeoutId);
			this.#timeoutId = void 0;
		}
	};
};
/**
* Creates a debounced function that delays invoking the provided function until after a specified wait time.
* Multiple calls during the wait period will cancel previous pending invocations and reset the timer.
*
* This synchronous version is lighter weight and often all you need - upgrade to asyncDebounce when you need promises, retry support, abort/cancel capabilities, or advanced error handling.
*
* If leading option is true, the function will execute immediately on the first call, then wait the delay
* before allowing another execution.
*
* State Management:
* - Uses TanStack Store for reactive state management
* - Use `initialState` to provide initial state values when creating the debouncer
* - Use `onExecute` callback to react to function execution and implement custom logic
* - The state includes canLeadingExecute, execution count, and isPending status
* - State can be accessed via the underlying Debouncer instance's `store.state` property
* - When using framework adapters (React/Solid), state is accessed from the hook's state property
*
* @example
* ```ts
* const debounced = debounce(() => {
*   saveChanges();
* }, { wait: 1000 });
*
* // Called repeatedly but executes at most once per second
* inputElement.addEventListener('input', debounced);
* ```
*/
function debounce(fn, initialOptions) {
	return new Debouncer(fn, initialOptions).maybeExecute;
}

//#endregion
export { Debouncer, debounce, debouncerOptions };
