// @tanstack/pacer/dist/event-client.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
let _tanstack_devtools_event_client = require("@tanstack/devtools-event-client");

//#region src/event-client.ts
const pacerDevtoolsInstancesByKey = /* @__PURE__ */ new Map();
function registerPacerDevtoolsInstance(key, instance) {
	pacerDevtoolsInstancesByKey.set(key, instance);
}
function getPacerDevtoolsInstance(key) {
	return pacerDevtoolsInstancesByKey.get(key);
}
function cloneJsonSafe(value) {
	if (value === void 0) return;
	try {
		return JSON.parse(JSON.stringify(value));
	} catch {
		return null;
	}
}
function readStoreSnapshot(store) {
	if (typeof store.get === "function") return store.get();
	return store.state;
}
function toPacerDevtoolsWirePayload(instance) {
	return {
		key: instance.key,
		store: { state: cloneJsonSafe(readStoreSnapshot(instance.store)) },
		options: cloneJsonSafe(instance.options)
	};
}
var PacerEventClient = class extends _tanstack_devtools_event_client.EventClient {
	constructor(props) {
		super({
			pluginId: "pacer",
			debug: props?.debug,
			reconnectEveryMs: 1e3
		});
	}
};
const emitChange = (event, instance) => {
	const key = instance.key;
	if (!key) return;
	registerPacerDevtoolsInstance(key, instance);
	pacerEventClient.emit(event, toPacerDevtoolsWirePayload({
		...instance,
		key
	}));
};
const pacerEventClient = new PacerEventClient();

//#endregion
exports.emitChange = emitChange;
exports.getPacerDevtoolsInstance = getPacerDevtoolsInstance;
exports.pacerEventClient = pacerEventClient;
exports.registerPacerDevtoolsInstance = registerPacerDevtoolsInstance;
