// @tanstack/pacer/dist/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_utils = require('./utils.cjs');
const require_event_client = require('./event-client.cjs');
const require_async_retryer = require('./async-retryer.cjs');
const require_async_batcher = require('./async-batcher.cjs');
const require_async_debouncer = require('./async-debouncer.cjs');
const require_async_queuer = require('./async-queuer.cjs');
const require_async_rate_limiter = require('./async-rate-limiter.cjs');
const require_async_throttler = require('./async-throttler.cjs');
const require_batcher = require('./batcher.cjs');
const require_debouncer = require('./debouncer.cjs');
const require_queuer = require('./queuer.cjs');
const require_rate_limiter = require('./rate-limiter.cjs');
const require_throttler = require('./throttler.cjs');

exports.AsyncBatcher = require_async_batcher.AsyncBatcher;
exports.AsyncDebouncer = require_async_debouncer.AsyncDebouncer;
exports.AsyncQueuer = require_async_queuer.AsyncQueuer;
exports.AsyncRateLimiter = require_async_rate_limiter.AsyncRateLimiter;
exports.AsyncRetryer = require_async_retryer.AsyncRetryer;
exports.AsyncThrottler = require_async_throttler.AsyncThrottler;
exports.Batcher = require_batcher.Batcher;
exports.Debouncer = require_debouncer.Debouncer;
exports.Queuer = require_queuer.Queuer;
exports.RateLimiter = require_rate_limiter.RateLimiter;
exports.Throttler = require_throttler.Throttler;
exports.asyncBatch = require_async_batcher.asyncBatch;
exports.asyncBatcherOptions = require_async_batcher.asyncBatcherOptions;
exports.asyncDebounce = require_async_debouncer.asyncDebounce;
exports.asyncDebouncerOptions = require_async_debouncer.asyncDebouncerOptions;
exports.asyncQueue = require_async_queuer.asyncQueue;
exports.asyncQueuerOptions = require_async_queuer.asyncQueuerOptions;
exports.asyncRateLimit = require_async_rate_limiter.asyncRateLimit;
exports.asyncRateLimiterOptions = require_async_rate_limiter.asyncRateLimiterOptions;
exports.asyncRetry = require_async_retryer.asyncRetry;
exports.asyncRetryerOptions = require_async_retryer.asyncRetryerOptions;
exports.asyncThrottle = require_async_throttler.asyncThrottle;
exports.asyncThrottlerOptions = require_async_throttler.asyncThrottlerOptions;
exports.batch = require_batcher.batch;
exports.debounce = require_debouncer.debounce;
exports.debouncerOptions = require_debouncer.debouncerOptions;
exports.emitChange = require_event_client.emitChange;
exports.getPacerDevtoolsInstance = require_event_client.getPacerDevtoolsInstance;
exports.isFunction = require_utils.isFunction;
exports.pacerEventClient = require_event_client.pacerEventClient;
exports.parseFunctionOrValue = require_utils.parseFunctionOrValue;
exports.queue = require_queuer.queue;
exports.queuerOptions = require_queuer.queuerOptions;
exports.rateLimit = require_rate_limiter.rateLimit;
exports.rateLimiterOptions = require_rate_limiter.rateLimiterOptions;
exports.throttle = require_throttler.throttle;
exports.throttlerOptions = require_throttler.throttlerOptions;
