// @tanstack/pacer/dist/rate-limiter.js
import { parseFunctionOrValue } from "./utils.js";
import { emitChange, pacerEventClient } from "./event-client.js";
import { Store } from "@tanstack/store";

//#region src/rate-limiter.ts
function getDefaultRateLimiterState() {
	return {
		executionCount: 0,
		executionTimes: [],
		isExceeded: false,
		rejectionCount: 0,
		status: "idle",
		maybeExecuteCount: 0
	};
}
/**
* Utility function for sharing common `RateLimiterOptions` options between different `RateLimiter` instances.
*/
function rateLimiterOptions(options) {
	return options;
}
const defaultOptions = {
	enabled: true,
	limit: 1,
	window: 0,
	windowType: "fixed"
};
/**
* A class that creates a rate-limited function.
*
* Rate limiting is a simple approach that allows a function to execute up to a limit within a time window,
* then blocks all subsequent calls until the window passes. This can lead to "bursty" behavior where
* all executions happen immediately, followed by a complete block.
* This synchronous version is lighter weight and often all you need - upgrade to AsyncRateLimiter when you need promises, retry support, abort capabilities, or advanced error handling.
*
* The rate limiter supports two types of windows:
* - 'fixed': A strict window that resets after the window period. All executions within the window count
*   towards the limit, and the window resets completely after the period.
* - 'sliding': A rolling window that allows executions as old ones expire. This provides a more
*   consistent rate of execution over time.
*
* For smoother execution patterns, consider using:
* - Throttling: Ensures consistent spacing between executions (e.g. max once per 200ms)
* - Debouncing: Waits for a pause in calls before executing (e.g. after 500ms of no calls)
*
* Rate limiting is best used for hard API limits or resource constraints. For UI updates or
* smoothing out frequent events, throttling or debouncing usually provide better user experience.
*
* State Management:
* - Uses TanStack Store for reactive state management
* - Use `initialState` to provide initial state values when creating the rate limiter
* - Use `onExecute` callback to react to function execution and implement custom logic
* - Use `onReject` callback to react to executions being rejected when rate limit is exceeded
* - The state includes execution count, execution times, and rejection count
* - State can be accessed via `rateLimiter.store.state` when using the class directly
* - When using framework adapters (React/Solid), state is accessed from `rateLimiter.state`
*
* @example
* ```ts
* const rateLimiter = new RateLimiter(
*   (id: string) => api.getData(id),
*   {
*     limit: 5,
*     window: 1000,
*     windowType: 'sliding',
*   }
* );
*
* // Will execute immediately until limit reached, then block
* rateLimiter.maybeExecute('123');
* ```
*/
var RateLimiter = class {
	#timeoutIds = /* @__PURE__ */ new Set();
	constructor(fn, initialOptions) {
		this.fn = fn;
		this.store = new Store(getDefaultRateLimiterState());
		this.setOptions = (newOptions) => {
			this.options = {
				...this.options,
				...newOptions
			};
		};
		this.maybeExecute = (...args) => {
			this.#setState({ maybeExecuteCount: this.store.state.maybeExecuteCount + 1 });
			this.#cleanupOldExecutions();
			if (this.#getExecutionTimesInWindow().length < this.#getLimit()) {
				this.#execute(...args);
				return true;
			}
			this.#setState({ rejectionCount: this.store.state.rejectionCount + 1 });
			this.options.onReject?.(this);
			return false;
		};
		this.getRemainingInWindow = () => {
			const relevantExecutionTimes = this.#getExecutionTimesInWindow();
			return Math.max(0, this.#getLimit() - relevantExecutionTimes.length);
		};
		this.getMsUntilNextWindow = () => {
			if (this.getRemainingInWindow() > 0) return 0;
			return (this.store.state.executionTimes[0] ?? Infinity) + this.#getWindow() - Date.now();
		};
		this.reset = () => {
			this.#setState(getDefaultRateLimiterState());
			this.#clearTimeouts();
		};
		this.key = initialOptions.key;
		this.options = {
			...defaultOptions,
			...initialOptions
		};
		this.#setState(this.options.initialState ?? {});
		for (const executionTime of this.#getExecutionTimesInWindow()) this.#setCleanupTimeout(executionTime);
		if (this.key) pacerEventClient.on("d-RateLimiter", (event) => {
			if (event.payload.key !== this.key) return;
			this.#setState(event.payload.store.state);
			this.setOptions(event.payload.options);
		});
	}
	#setState = (newState) => {
		this.store.setState((state) => {
			const combinedState = {
				...state,
				...newState
			};
			const isExceeded = combinedState.executionTimes.length >= this.#getLimit();
			const status = !this.#getEnabled() ? "disabled" : isExceeded ? "exceeded" : "idle";
			return {
				...combinedState,
				isExceeded,
				status
			};
		});
		emitChange("RateLimiter", this);
	};
	/**
	* Returns the current enabled state of the rate limiter
	*/
	#getEnabled = () => {
		return !!parseFunctionOrValue(this.options.enabled, this);
	};
	/**
	* Returns the current limit of executions allowed within the time window
	*/
	#getLimit = () => {
		return parseFunctionOrValue(this.options.limit, this);
	};
	/**
	* Returns the current time window in milliseconds
	*/
	#getWindow = () => {
		return parseFunctionOrValue(this.options.window, this);
	};
	#execute = (...args) => {
		if (!this.#getEnabled()) return;
		const now = Date.now();
		this.fn(...args);
		this.store.state.executionTimes.push(now);
		this.#setCleanupTimeout(now);
		this.#setState({ executionCount: this.store.state.executionCount + 1 });
		this.options.onExecute?.(args, this);
	};
	#getExecutionTimesInWindow = () => {
		if (this.options.windowType === "sliding") return this.store.state.executionTimes.filter((time) => time > Date.now() - this.#getWindow());
		else {
			if (this.store.state.executionTimes.length === 0) return [];
			const windowStart = Math.min(...this.store.state.executionTimes);
			const windowEnd = windowStart + this.#getWindow();
			if (Date.now() > windowEnd) return [];
			return this.store.state.executionTimes.filter((time) => time >= windowStart && time <= windowEnd);
		}
	};
	#setCleanupTimeout = (executionTime) => {
		if (this.options.windowType === "sliding" || this.#timeoutIds.size === 0) {
			const timeUntilExpiration = executionTime - Date.now() + this.#getWindow() + 1;
			const timeoutId = setTimeout(() => {
				this.#cleanupOldExecutions();
				this.#clearTimeout(timeoutId);
			}, timeUntilExpiration);
			this.#timeoutIds.add(timeoutId);
		}
	};
	#clearTimeout = (timeoutId) => {
		clearTimeout(timeoutId);
		this.#timeoutIds.delete(timeoutId);
	};
	#clearTimeouts = () => {
		this.#timeoutIds.forEach((timeoutId) => clearTimeout(timeoutId));
		this.#timeoutIds.clear();
	};
	#cleanupOldExecutions = () => {
		this.#setState({ executionTimes: this.#getExecutionTimesInWindow() });
	};
};
/**
* Creates a rate-limited function that will execute the provided function up to a maximum number of times within a time window.
*
* This synchronous version is lighter weight and often all you need - upgrade to asyncRateLimit when you need promises, retry support, abort capabilities, or advanced error handling.
*
* Note that rate limiting is a simpler form of execution control compared to throttling or debouncing:
* - A rate limiter will allow all executions until the limit is reached, then block all subsequent calls until the window resets
* - A throttler ensures even spacing between executions, which can be better for consistent performance
* - A debouncer collapses multiple calls into one, which is better for handling bursts of events
*
* The rate limiter supports two types of windows:
* - 'fixed': A strict window that resets after the window period. All executions within the window count
*   towards the limit, and the window resets completely after the period.
* - 'sliding': A rolling window that allows executions as old ones expire. This provides a more
*   consistent rate of execution over time.
*
* State Management:
* - Uses TanStack Store for reactive state management
* - Use `initialState` to provide initial state values when creating the rate limiter
* - Use `onExecute` callback to react to function execution and implement custom logic
* - Use `onReject` callback to react to executions being rejected when rate limit is exceeded
* - The state includes execution count, execution times, and rejection count
* - State can be accessed via the underlying RateLimiter instance's `store.state` property
* - When using framework adapters (React/Solid), state is accessed from the hook's state property
*
* Consider using throttle() or debounce() if you need more intelligent execution control. Use rate limiting when you specifically
* need to enforce a hard limit on the number of executions within a time period.
*
* @example
* ```ts
* // Rate limit to 5 calls per minute with a sliding window
* const rateLimited = rateLimit(makeApiCall, {
*   limit: 5,
*   window: 60000,
*   windowType: 'sliding',
*   onReject: (rateLimiter) => {
*     console.log(`Rate limit exceeded. Try again in ${rateLimiter.getMsUntilNextWindow()}ms`);
*   }
* });
*
* // First 5 calls will execute immediately
* // Additional calls will be rejected until the minute window resets
* rateLimited();
*
* // For more even execution, consider using throttle instead:
* const throttled = throttle(makeApiCall, { wait: 12000 }); // One call every 12 seconds
* ```
*/
function rateLimit(fn, initialOptions) {
	return new RateLimiter(fn, initialOptions).maybeExecute;
}

//#endregion
export { RateLimiter, rateLimit, rateLimiterOptions };
