// @tanstack/pacer/dist/throttler.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_utils = require('./utils.cjs');
const require_event_client = require('./event-client.cjs');
let _tanstack_store = require("@tanstack/store");

//#region src/throttler.ts
function getDefaultThrottlerState() {
	return {
		executionCount: 0,
		isPending: false,
		lastArgs: void 0,
		lastExecutionTime: 0,
		nextExecutionTime: 0,
		status: "idle",
		maybeExecuteCount: 0
	};
}
/**
* Utility function for sharing common `ThrottlerOptions` options between different `Throttler` instances.
*/
function throttlerOptions(options) {
	return options;
}
const defaultOptions = {
	enabled: true,
	leading: true,
	trailing: true,
	wait: 0
};
/**
* A class that creates a throttled function.
*
* Throttling ensures a function is called at most once within a specified time window.
* Unlike debouncing which waits for a pause in calls, throttling guarantees consistent
* execution timing regardless of call frequency.
* This synchronous version is lighter weight and often all you need - upgrade to AsyncThrottler when you need promises, retry support, abort/cancel capabilities, or advanced error handling.
*
* Supports both leading and trailing edge execution:
* - Leading: Execute immediately on first call (default: true)
* - Trailing: Execute after wait period if called during throttle (default: true)
*
* For collapsing rapid-fire events where you only care about the last call, consider using Debouncer.
*
* State Management:
* - Uses TanStack Store for reactive state management
* - Use `initialState` to provide initial state values when creating the throttler
* - Use `onExecute` callback to react to function execution and implement custom logic
* - The state includes execution count, last execution time, pending status, and more
* - State can be accessed via `throttler.store.state` when using the class directly
* - When using framework adapters (React/Solid), state is accessed from `throttler.state`
*
* @example
* ```ts
* const throttler = new Throttler(
*   (id: string) => api.getData(id),
*   { wait: 1000 } // Execute at most once per second
* );
*
* // First call executes immediately
* throttler.maybeExecute('123');
*
* // Subsequent calls within 1000ms are throttled
* throttler.maybeExecute('123'); // Throttled
* ```
*/
var Throttler = class {
	#timeoutId;
	constructor(fn, initialOptions) {
		this.fn = fn;
		this.store = new _tanstack_store.Store(getDefaultThrottlerState());
		this.setOptions = (newOptions) => {
			this.options = {
				...this.options,
				...newOptions
			};
			if (!this.#getEnabled()) this.cancel();
		};
		this.maybeExecute = (...args) => {
			this.#setState({ maybeExecuteCount: this.store.state.maybeExecuteCount + 1 });
			const now = Date.now();
			const timeSinceLastExecution = now - this.store.state.lastExecutionTime;
			const wait = this.#getWait();
			if (this.options.leading && timeSinceLastExecution >= wait) this.#execute(...args);
			else {
				this.#setState({ lastArgs: args });
				if (!this.#timeoutId && this.options.trailing) {
					const timeoutDuration = wait - (this.store.state.lastExecutionTime ? now - this.store.state.lastExecutionTime : 0);
					this.#setState({ isPending: true });
					this.#timeoutId = setTimeout(() => {
						const { lastArgs } = this.store.state;
						if (lastArgs !== void 0) this.#execute(...lastArgs);
					}, timeoutDuration);
				}
			}
		};
		this.flush = () => {
			if (this.store.state.isPending && this.store.state.lastArgs) this.#execute(...this.store.state.lastArgs);
		};
		this.cancel = () => {
			this.#clearTimeout();
			this.#setState({
				lastArgs: void 0,
				isPending: false
			});
		};
		this.reset = () => {
			this.#setState(getDefaultThrottlerState());
		};
		this.key = initialOptions.key;
		this.options = {
			...defaultOptions,
			...initialOptions
		};
		this.#setState(this.options.initialState ?? {});
		if (this.key) require_event_client.pacerEventClient.on("d-Throttler", (event) => {
			if (event.payload.key !== this.key) return;
			this.#setState(event.payload.store.state);
			this.setOptions(event.payload.options);
		});
	}
	#setState = (newState) => {
		this.store.setState((state) => {
			const combinedState = {
				...state,
				...newState
			};
			const { isPending } = combinedState;
			return {
				...combinedState,
				status: !this.#getEnabled() ? "disabled" : isPending ? "pending" : "idle"
			};
		});
		require_event_client.emitChange("Throttler", this);
	};
	#getEnabled = () => {
		return !!require_utils.parseFunctionOrValue(this.options.enabled, this);
	};
	#getWait = () => {
		return require_utils.parseFunctionOrValue(this.options.wait, this);
	};
	#execute = (...args) => {
		if (!this.#getEnabled()) return;
		this.fn(...args);
		const lastExecutionTime = Date.now();
		const nextExecutionTime = lastExecutionTime + this.#getWait();
		this.#clearTimeout();
		this.#setState({
			executionCount: this.store.state.executionCount + 1,
			lastExecutionTime,
			nextExecutionTime,
			isPending: false,
			lastArgs: void 0
		});
		this.options.onExecute?.(args, this);
		setTimeout(() => {
			if (!this.store.state.isPending) this.#setState({ nextExecutionTime: void 0 });
		}, this.#getWait());
	};
	#clearTimeout = () => {
		if (this.#timeoutId) {
			clearTimeout(this.#timeoutId);
			this.#timeoutId = void 0;
		}
	};
};
/**
* Creates a throttled function that limits how often the provided function can execute.
*
* This synchronous version is lighter weight and often all you need - upgrade to asyncThrottle when you need promises, retry support, abort/cancel capabilities, or advanced error handling.
*
* Throttling ensures a function executes at most once within a specified time window,
* regardless of how many times it is called. This is useful for rate-limiting
* expensive operations or UI updates.
*
* The throttled function can be configured to execute on the leading and/or trailing
* edge of the throttle window via options.
*
* For handling bursts of events, consider using debounce() instead. For hard execution
* limits, consider using rateLimit().
*
* State Management:
* - Uses TanStack Store for reactive state management
* - Use `initialState` to provide initial state values when creating the throttler
* - Use `onExecute` callback to react to function execution and implement custom logic
* - The state includes execution count, last execution time, pending status, and more
* - State can be accessed via the underlying Throttler instance's `store.state` property
* - When using framework adapters (React/Solid), state is accessed from the hook's state property
*
* @example
* ```ts
* // Basic throttling - max once per second
* const throttled = throttle(updateUI, { wait: 1000 });
*
* // Configure leading/trailing execution
* const throttled = throttle(saveData, {
*   wait: 2000,
*   leading: true,  // Execute immediately on first call
*   trailing: true  // Execute again after delay if called during wait
* });
* ```
*/
function throttle(fn, initialOptions) {
	return new Throttler(fn, initialOptions).maybeExecute;
}

//#endregion
exports.Throttler = Throttler;
exports.throttle = throttle;
exports.throttlerOptions = throttlerOptions;
